// @copilotkit/react-core/dist/context/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/context/index.ts
var context_exports = {};
__export(context_exports, {
  CoAgentStateRendersContext: () => CoAgentStateRendersContext,
  CoAgentStateRendersProvider: () => CoAgentStateRendersProvider,
  CopilotContext: () => CopilotContext,
  CopilotMessagesContext: () => CopilotMessagesContext,
  ThreadsContext: () => ThreadsContext,
  ThreadsProvider: () => ThreadsProvider,
  useCoAgentStateRenders: () => useCoAgentStateRenders,
  useCopilotContext: () => useCopilotContext,
  useCopilotMessagesContext: () => useCopilotMessagesContext,
  useThreads: () => useThreads
});
module.exports = __toCommonJS(context_exports);

// src/context/copilot-context.tsx
var import_react = __toESM(require("react"));
var emptyCopilotContext = {
  actions: {},
  setAction: () => {
  },
  removeAction: () => {
  },
  setRegisteredActions: () => "",
  removeRegisteredAction: () => {
  },
  chatComponentsCache: { current: { actions: {}, coAgentStateRenders: {} } },
  getContextString: (documents, categories) => returnAndThrowInDebug(""),
  addContext: () => "",
  removeContext: () => {
  },
  getAllContext: () => [],
  getFunctionCallHandler: () => returnAndThrowInDebug(() => __async(void 0, null, function* () {
  })),
  isLoading: false,
  setIsLoading: () => returnAndThrowInDebug(false),
  chatInstructions: "",
  setChatInstructions: () => returnAndThrowInDebug(""),
  additionalInstructions: [],
  setAdditionalInstructions: () => returnAndThrowInDebug([]),
  getDocumentsContext: (categories) => returnAndThrowInDebug([]),
  addDocumentContext: () => returnAndThrowInDebug(""),
  removeDocumentContext: () => {
  },
  copilotApiConfig: new class {
    get chatApiEndpoint() {
      throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
    }
    get headers() {
      return {};
    }
    get body() {
      return {};
    }
  }(),
  chatSuggestionConfiguration: {},
  addChatSuggestionConfiguration: () => {
  },
  removeChatSuggestionConfiguration: () => {
  },
  showDevConsole: false,
  coagentStates: {},
  setCoagentStates: () => {
  },
  coagentStatesRef: { current: {} },
  setCoagentStatesWithRef: () => {
  },
  agentSession: null,
  setAgentSession: () => {
  },
  forwardedParameters: {},
  agentLock: null,
  threadId: "",
  setThreadId: () => {
  },
  runId: null,
  setRunId: () => {
  },
  chatAbortControllerRef: { current: null },
  availableAgents: [],
  extensions: {},
  setExtensions: () => {
  },
  interruptActions: {},
  setInterruptAction: () => {
  },
  removeInterruptAction: () => {
  },
  interruptEventQueue: {},
  addInterruptEvent: () => {
  },
  resolveInterruptEvent: () => {
  },
  onError: () => {
  },
  bannerError: null,
  setBannerError: () => {
  },
  internalErrorHandlers: {},
  setInternalErrorHandler: () => {
  },
  removeInternalErrorHandler: () => {
  }
};
var CopilotContext = import_react.default.createContext(emptyCopilotContext);
function useCopilotContext() {
  const context = import_react.default.useContext(CopilotContext);
  if (context === emptyCopilotContext) {
    throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
  }
  return context;
}
function returnAndThrowInDebug(_value) {
  throw new Error("Remember to wrap your app in a `<CopilotKit> {...} </CopilotKit>` !!!");
}

// src/context/copilot-messages-context.tsx
var import_react2 = __toESM(require("react"));
var emptyCopilotContext2 = {
  messages: [],
  setMessages: () => [],
  // suggestions state
  suggestions: [],
  setSuggestions: () => []
};
var CopilotMessagesContext = import_react2.default.createContext(emptyCopilotContext2);
function useCopilotMessagesContext() {
  const context = import_react2.default.useContext(CopilotMessagesContext);
  if (context === emptyCopilotContext2) {
    throw new Error(
      "A messages consuming component was not wrapped with `<CopilotMessages> {...} </CopilotMessages>`"
    );
  }
  return context;
}

// src/context/coagent-state-renders-context.tsx
var import_react3 = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var CoAgentStateRendersContext = (0, import_react3.createContext)(
  void 0
);
function CoAgentStateRendersProvider({ children }) {
  const [coAgentStateRenders, setCoAgentStateRenders] = (0, import_react3.useState)({});
  const setCoAgentStateRender = (0, import_react3.useCallback)((id, stateRender) => {
    setCoAgentStateRenders((prevPoints) => __spreadProps(__spreadValues({}, prevPoints), {
      [id]: stateRender
    }));
  }, []);
  const removeCoAgentStateRender = (0, import_react3.useCallback)((id) => {
    setCoAgentStateRenders((prevPoints) => {
      const newPoints = __spreadValues({}, prevPoints);
      delete newPoints[id];
      return newPoints;
    });
  }, []);
  const claimsRef = (0, import_react3.useRef)({});
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    CoAgentStateRendersContext.Provider,
    {
      value: {
        coAgentStateRenders,
        setCoAgentStateRender,
        removeCoAgentStateRender,
        claimsRef
      },
      children
    }
  );
}
function useCoAgentStateRenders() {
  const context = (0, import_react3.useContext)(CoAgentStateRendersContext);
  if (!context) {
    throw new Error("useCoAgentStateRenders must be used within CoAgentStateRendersProvider");
  }
  return context;
}

// src/context/threads-context.tsx
var import_react4 = require("react");
var import_shared = require("@copilotkit/shared");
var import_jsx_runtime2 = require("react/jsx-runtime");
var ThreadsContext = (0, import_react4.createContext)(void 0);
function ThreadsProvider({ children, threadId: explicitThreadId }) {
  const [internalThreadId, setThreadId] = (0, import_react4.useState)(() => (0, import_shared.randomUUID)());
  const threadId = explicitThreadId != null ? explicitThreadId : internalThreadId;
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    ThreadsContext.Provider,
    {
      value: {
        threadId,
        setThreadId
      },
      children
    }
  );
}
function useThreads() {
  const context = (0, import_react4.useContext)(ThreadsContext);
  if (!context) {
    throw new Error("useThreads must be used within ThreadsProvider");
  }
  return context;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CoAgentStateRendersContext,
  CoAgentStateRendersProvider,
  CopilotContext,
  CopilotMessagesContext,
  ThreadsContext,
  ThreadsProvider,
  useCoAgentStateRenders,
  useCopilotContext,
  useCopilotMessagesContext,
  useThreads
});
