// @copilotkitnext/web-inspector/dist/index.mjs
// src/index.ts
import { LitElement, css, html, nothing, unsafeCSS } from "lit";
import { styleMap } from "lit/directives/style-map.js";

// src/styles/generated.css
var generated_default = '/*! tailwindcss v4.1.18 | MIT License | https://tailwindcss.com */\n@layer properties{@supports (((-webkit-hyphens:none)) and (not (margin-trim:inline))) or ((-moz-orient:inline) and (not (color:rgb(from red r g b)))){*,:before,:after,::backdrop{--tw-translate-x:0;--tw-translate-y:0;--tw-translate-z:0;--tw-rotate-x:initial;--tw-rotate-y:initial;--tw-rotate-z:initial;--tw-skew-x:initial;--tw-skew-y:initial;--tw-space-y-reverse:0;--tw-divide-y-reverse:0;--tw-border-style:solid;--tw-leading:initial;--tw-font-weight:initial;--tw-shadow:0 0 #0000;--tw-shadow-color:initial;--tw-shadow-alpha:100%;--tw-inset-shadow:0 0 #0000;--tw-inset-shadow-color:initial;--tw-inset-shadow-alpha:100%;--tw-ring-color:initial;--tw-ring-shadow:0 0 #0000;--tw-inset-ring-color:initial;--tw-inset-ring-shadow:0 0 #0000;--tw-ring-inset:initial;--tw-ring-offset-width:0px;--tw-ring-offset-color:#fff;--tw-ring-offset-shadow:0 0 #0000;--tw-backdrop-blur:initial;--tw-backdrop-brightness:initial;--tw-backdrop-contrast:initial;--tw-backdrop-grayscale:initial;--tw-backdrop-hue-rotate:initial;--tw-backdrop-invert:initial;--tw-backdrop-opacity:initial;--tw-backdrop-saturate:initial;--tw-backdrop-sepia:initial;--tw-scale-x:1;--tw-scale-y:1;--tw-scale-z:1;--tw-outline-style:solid}}}@layer theme{:root,:host{--font-sans:ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";--font-mono:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;--color-amber-50:oklch(98.7% .022 95.277);--color-amber-100:oklch(96.2% .059 95.617);--color-amber-200:oklch(92.4% .12 95.746);--color-amber-600:oklch(66.6% .179 58.318);--color-amber-700:oklch(55.5% .163 48.998);--color-amber-800:oklch(47.3% .137 46.201);--color-amber-900:oklch(41.4% .112 45.904);--color-green-100:oklch(96.2% .044 156.743);--color-green-700:oklch(52.7% .154 150.069);--color-green-800:oklch(44.8% .119 151.328);--color-emerald-50:oklch(97.9% .021 166.113);--color-emerald-200:oklch(90.5% .093 164.15);--color-emerald-500:oklch(69.6% .17 162.48);--color-emerald-700:oklch(50.8% .118 165.612);--color-sky-50:oklch(97.7% .013 236.62);--color-sky-200:oklch(90.1% .058 230.902);--color-sky-700:oklch(50% .134 242.749);--color-blue-50:oklch(97% .014 254.604);--color-blue-100:oklch(93.2% .032 255.585);--color-blue-200:oklch(88.2% .059 254.128);--color-blue-600:oklch(54.6% .245 262.881);--color-blue-700:oklch(48.8% .243 264.376);--color-blue-800:oklch(42.4% .199 265.638);--color-violet-50:oklch(96.9% .016 293.756);--color-violet-200:oklch(89.4% .057 293.283);--color-violet-700:oklch(49.1% .27 292.581);--color-purple-50:oklch(97.7% .014 308.299);--color-purple-200:oklch(90.2% .063 306.703);--color-purple-700:oklch(49.6% .265 301.924);--color-rose-50:oklch(96.9% .015 12.422);--color-rose-200:oklch(89.2% .058 10.001);--color-rose-500:oklch(64.5% .246 16.439);--color-rose-600:oklch(58.6% .253 17.585);--color-rose-700:oklch(51.4% .222 16.935);--color-rose-800:oklch(45.5% .188 13.697);--color-rose-900:oklch(41% .159 10.272);--color-slate-200:oklch(92.9% .013 255.508);--color-slate-800:oklch(27.9% .041 260.031);--color-slate-900:oklch(20.8% .042 265.755);--color-slate-950:oklch(12.9% .042 264.695);--color-gray-50:oklch(98.5% .002 247.839);--color-gray-100:oklch(96.7% .003 264.542);--color-gray-200:oklch(92.8% .006 264.531);--color-gray-300:oklch(87.2% .01 258.338);--color-gray-400:oklch(70.7% .022 261.325);--color-gray-500:oklch(55.1% .027 264.364);--color-gray-600:oklch(44.6% .03 256.802);--color-gray-700:oklch(37.3% .034 259.733);--color-gray-800:oklch(27.8% .033 256.848);--color-gray-900:oklch(21% .034 264.665);--color-black:#000;--color-white:#fff;--spacing:.25rem;--container-xs:20rem;--container-md:28rem;--container-2xl:42rem;--text-xs:.75rem;--text-xs--line-height:calc(1/.75);--text-sm:.875rem;--text-sm--line-height:calc(1.25/.875);--text-lg:1.125rem;--text-lg--line-height:calc(1.75/1.125);--font-weight-medium:500;--font-weight-semibold:600;--leading-snug:1.375;--leading-relaxed:1.625;--radius-sm:.25rem;--radius-md:.375rem;--radius-lg:.5rem;--radius-xl:.75rem;--animate-pulse:pulse 2s cubic-bezier(.4,0,.6,1)infinite;--blur-sm:8px;--blur-md:12px;--default-transition-duration:.15s;--default-transition-timing-function:cubic-bezier(.4,0,.2,1);--default-font-family:var(--font-sans);--default-mono-font-family:var(--font-mono)}}@layer base{*,:after,:before,::backdrop{box-sizing:border-box;border:0 solid;margin:0;padding:0}::file-selector-button{box-sizing:border-box;border:0 solid;margin:0;padding:0}html,:host{-webkit-text-size-adjust:100%;tab-size:4;line-height:1.5;font-family:var(--default-font-family,ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji");font-feature-settings:var(--default-font-feature-settings,normal);font-variation-settings:var(--default-font-variation-settings,normal);-webkit-tap-highlight-color:transparent}hr{height:0;color:inherit;border-top-width:1px}abbr:where([title]){-webkit-text-decoration:underline dotted;text-decoration:underline dotted}h1,h2,h3,h4,h5,h6{font-size:inherit;font-weight:inherit}a{color:inherit;-webkit-text-decoration:inherit;-webkit-text-decoration:inherit;-webkit-text-decoration:inherit;text-decoration:inherit}b,strong{font-weight:bolder}code,kbd,samp,pre{font-family:var(--default-mono-font-family,ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace);font-feature-settings:var(--default-mono-font-feature-settings,normal);font-variation-settings:var(--default-mono-font-variation-settings,normal);font-size:1em}small{font-size:80%}sub,sup{vertical-align:baseline;font-size:75%;line-height:0;position:relative}sub{bottom:-.25em}sup{top:-.5em}table{text-indent:0;border-color:inherit;border-collapse:collapse}:-moz-focusring{outline:auto}progress{vertical-align:baseline}summary{display:list-item}ol,ul,menu{list-style:none}img,svg,video,canvas,audio,iframe,embed,object{vertical-align:middle;display:block}img,video{max-width:100%;height:auto}button,input,select,optgroup,textarea{font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;color:inherit;opacity:1;background-color:#0000;border-radius:0}::file-selector-button{font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;color:inherit;opacity:1;background-color:#0000;border-radius:0}:where(select:is([multiple],[size])) optgroup{font-weight:bolder}:where(select:is([multiple],[size])) optgroup option{padding-inline-start:20px}::file-selector-button{margin-inline-end:4px}::placeholder{opacity:1}@supports (not ((-webkit-appearance:-apple-pay-button))) or (contain-intrinsic-size:1px){::placeholder{color:currentColor}@supports (color:color-mix(in lab, red, red)){::placeholder{color:color-mix(in oklab,currentcolor 50%,transparent)}}}textarea{resize:vertical}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-date-and-time-value{min-height:1lh;text-align:inherit}::-webkit-datetime-edit{display:inline-flex}::-webkit-datetime-edit-fields-wrapper{padding:0}::-webkit-datetime-edit{padding-block:0}::-webkit-datetime-edit-year-field{padding-block:0}::-webkit-datetime-edit-month-field{padding-block:0}::-webkit-datetime-edit-day-field{padding-block:0}::-webkit-datetime-edit-hour-field{padding-block:0}::-webkit-datetime-edit-minute-field{padding-block:0}::-webkit-datetime-edit-second-field{padding-block:0}::-webkit-datetime-edit-millisecond-field{padding-block:0}::-webkit-datetime-edit-meridiem-field{padding-block:0}::-webkit-calendar-picker-indicator{line-height:1}:-moz-ui-invalid{box-shadow:none}button,input:where([type=button],[type=reset],[type=submit]){appearance:button}::file-selector-button{appearance:button}::-webkit-inner-spin-button{height:auto}::-webkit-outer-spin-button{height:auto}[hidden]:where(:not([hidden=until-found])){display:none!important}:host{font-family:var(--font-sans);color:var(--color-slate-900);background-color:#0000;display:block}}@layer components;@layer utilities{.pointer-events-auto{pointer-events:auto}.absolute{position:absolute}.fixed{position:fixed}.relative{position:relative}.static{position:static}.sticky{position:sticky}.top-0{top:calc(var(--spacing)*0)}.right-0{right:calc(var(--spacing)*0)}.right-1{right:calc(var(--spacing)*1)}.bottom-1{bottom:calc(var(--spacing)*1)}.left-0{left:calc(var(--spacing)*0)}.z-10{z-index:10}.z-30{z-index:30}.z-40{z-index:40}.z-50{z-index:50}.m-0{margin:calc(var(--spacing)*0)}.mx-4{margin-inline:calc(var(--spacing)*4)}.my-1{margin-block:calc(var(--spacing)*1)}.my-3{margin-block:calc(var(--spacing)*3)}.mt-0\\.5{margin-top:calc(var(--spacing)*.5)}.mt-1{margin-top:calc(var(--spacing)*1)}.mt-1\\.5{margin-top:calc(var(--spacing)*1.5)}.mt-2{margin-top:calc(var(--spacing)*2)}.mb-1{margin-bottom:calc(var(--spacing)*1)}.mb-2{margin-bottom:calc(var(--spacing)*2)}.mb-3{margin-bottom:calc(var(--spacing)*3)}.mb-4{margin-bottom:calc(var(--spacing)*4)}.ml-auto{margin-left:auto}.box-border{box-sizing:border-box}.block{display:block}.flex{display:flex}.grid{display:grid}.inline-block{display:inline-block}.inline-flex{display:inline-flex}.h-1\\.5{height:calc(var(--spacing)*1.5)}.h-3{height:calc(var(--spacing)*3)}.h-3\\.5{height:calc(var(--spacing)*3.5)}.h-5{height:calc(var(--spacing)*5)}.h-6{height:calc(var(--spacing)*6)}.h-7{height:calc(var(--spacing)*7)}.h-8{height:calc(var(--spacing)*8)}.h-10{height:calc(var(--spacing)*10)}.h-12{height:calc(var(--spacing)*12)}.h-40{height:calc(var(--spacing)*40)}.h-full{height:100%}.max-h-64{max-height:calc(var(--spacing)*64)}.max-h-96{max-height:calc(var(--spacing)*96)}.w-1\\.5{width:calc(var(--spacing)*1.5)}.w-3{width:calc(var(--spacing)*3)}.w-3\\.5{width:calc(var(--spacing)*3.5)}.w-5{width:calc(var(--spacing)*5)}.w-6{width:calc(var(--spacing)*6)}.w-7{width:calc(var(--spacing)*7)}.w-8{width:calc(var(--spacing)*8)}.w-10{width:calc(var(--spacing)*10)}.w-12{width:calc(var(--spacing)*12)}.w-40{width:calc(var(--spacing)*40)}.w-auto{width:auto}.w-full{width:100%}.max-w-2xl{max-width:var(--container-2xl)}.max-w-\\[240px\\]{max-width:240px}.max-w-md{max-width:var(--container-md)}.max-w-xs{max-width:var(--container-xs)}.min-w-0{min-width:calc(var(--spacing)*0)}.min-w-\\[160px\\]{min-width:160px}.min-w-\\[200px\\]{min-width:200px}.flex-1{flex:1}.shrink-0{flex-shrink:0}.table-fixed{table-layout:fixed}.border-collapse{border-collapse:collapse}.-translate-y-\\[2px\\]{--tw-translate-y:calc(2px*-1);translate:var(--tw-translate-x)var(--tw-translate-y)}.rotate-180{rotate:180deg}.transform{transform:var(--tw-rotate-x,)var(--tw-rotate-y,)var(--tw-rotate-z,)var(--tw-skew-x,)var(--tw-skew-y,)}.animate-pulse{animation:var(--animate-pulse)}.cursor-grab{cursor:grab}.cursor-grabbing{cursor:grabbing}.cursor-nwse-resize{cursor:nwse-resize}.cursor-pointer{cursor:pointer}.touch-none{touch-action:none}.resize{resize:both}.grid-cols-2{grid-template-columns:repeat(2,minmax(0,1fr))}.flex-col{flex-direction:column}.flex-wrap{flex-wrap:wrap}.items-center{align-items:center}.items-start{align-items:flex-start}.justify-between{justify-content:space-between}.justify-center{justify-content:center}.gap-1{gap:calc(var(--spacing)*1)}.gap-1\\.5{gap:calc(var(--spacing)*1.5)}.gap-2{gap:calc(var(--spacing)*2)}.gap-3{gap:calc(var(--spacing)*3)}.gap-4{gap:calc(var(--spacing)*4)}:where(.space-y-1>:not(:last-child)){--tw-space-y-reverse:0;margin-block-start:calc(calc(var(--spacing)*1)*var(--tw-space-y-reverse));margin-block-end:calc(calc(var(--spacing)*1)*calc(1 - var(--tw-space-y-reverse)))}:where(.space-y-2>:not(:last-child)){--tw-space-y-reverse:0;margin-block-start:calc(calc(var(--spacing)*2)*var(--tw-space-y-reverse));margin-block-end:calc(calc(var(--spacing)*2)*calc(1 - var(--tw-space-y-reverse)))}:where(.space-y-3>:not(:last-child)){--tw-space-y-reverse:0;margin-block-start:calc(calc(var(--spacing)*3)*var(--tw-space-y-reverse));margin-block-end:calc(calc(var(--spacing)*3)*calc(1 - var(--tw-space-y-reverse)))}:where(.divide-y>:not(:last-child)){--tw-divide-y-reverse:0;border-bottom-style:var(--tw-border-style);border-top-style:var(--tw-border-style);border-top-width:calc(1px*var(--tw-divide-y-reverse));border-bottom-width:calc(1px*calc(1 - var(--tw-divide-y-reverse)))}:where(.divide-gray-200>:not(:last-child)){border-color:var(--color-gray-200)}.truncate{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}.overflow-auto{overflow:auto}.overflow-hidden{overflow:hidden}.overflow-x-hidden{overflow-x:hidden}.overflow-y-auto{overflow-y:auto}.rounded{border-radius:.25rem}.rounded-full{border-radius:3.40282e38px}.rounded-lg{border-radius:var(--radius-lg)}.rounded-md{border-radius:var(--radius-md)}.rounded-sm{border-radius:var(--radius-sm)}.rounded-xl{border-radius:var(--radius-xl)}.border{border-style:var(--tw-border-style);border-width:1px}.border-t{border-top-style:var(--tw-border-style);border-top-width:1px}.border-r{border-right-style:var(--tw-border-style);border-right-width:1px}.border-b{border-bottom-style:var(--tw-border-style);border-bottom-width:1px}.border-l{border-left-style:var(--tw-border-style);border-left-width:1px}.border-dashed{--tw-border-style:dashed;border-style:dashed}.border-amber-200{border-color:var(--color-amber-200)}.border-blue-200{border-color:var(--color-blue-200)}.border-emerald-200{border-color:var(--color-emerald-200)}.border-gray-100{border-color:var(--color-gray-100)}.border-gray-200{border-color:var(--color-gray-200)}.border-purple-200{border-color:var(--color-purple-200)}.border-rose-200{border-color:var(--color-rose-200)}.border-sky-200{border-color:var(--color-sky-200)}.border-slate-200{border-color:var(--color-slate-200)}.border-violet-200{border-color:var(--color-violet-200)}.border-white\\/20{border-color:#fff3}@supports (color:color-mix(in lab, red, red)){.border-white\\/20{border-color:color-mix(in oklab,var(--color-white)20%,transparent)}}.bg-amber-50{background-color:var(--color-amber-50)}.bg-amber-100{background-color:var(--color-amber-100)}.bg-blue-50{background-color:var(--color-blue-50)}.bg-blue-100{background-color:var(--color-blue-100)}.bg-emerald-50{background-color:var(--color-emerald-50)}.bg-emerald-500{background-color:var(--color-emerald-500)}.bg-gray-50{background-color:var(--color-gray-50)}.bg-gray-50\\/50{background-color:#f9fafb80}@supports (color:color-mix(in lab, red, red)){.bg-gray-50\\/50{background-color:color-mix(in oklab,var(--color-gray-50)50%,transparent)}}.bg-gray-100{background-color:var(--color-gray-100)}.bg-gray-400{background-color:var(--color-gray-400)}.bg-gray-900{background-color:var(--color-gray-900)}.bg-green-100{background-color:var(--color-green-100)}.bg-purple-50{background-color:var(--color-purple-50)}.bg-rose-50{background-color:var(--color-rose-50)}.bg-rose-500{background-color:var(--color-rose-500)}.bg-rose-600{background-color:var(--color-rose-600)}.bg-sky-50{background-color:var(--color-sky-50)}.bg-slate-900{background-color:var(--color-slate-900)}.bg-slate-950\\/95{background-color:#020618f2}@supports (color:color-mix(in lab, red, red)){.bg-slate-950\\/95{background-color:color-mix(in oklab,var(--color-slate-950)95%,transparent)}}.bg-violet-50{background-color:var(--color-violet-50)}.bg-white{background-color:var(--color-white)}.bg-white\\/60{background-color:#fff9}@supports (color:color-mix(in lab, red, red)){.bg-white\\/60{background-color:color-mix(in oklab,var(--color-white)60%,transparent)}}.bg-white\\/95{background-color:#fffffff2}@supports (color:color-mix(in lab, red, red)){.bg-white\\/95{background-color:color-mix(in oklab,var(--color-white)95%,transparent)}}.p-2{padding:calc(var(--spacing)*2)}.p-3{padding:calc(var(--spacing)*3)}.p-4{padding:calc(var(--spacing)*4)}.px-1{padding-inline:calc(var(--spacing)*1)}.px-1\\.5{padding-inline:calc(var(--spacing)*1.5)}.px-2{padding-inline:calc(var(--spacing)*2)}.px-3{padding-inline:calc(var(--spacing)*3)}.px-4{padding-inline:calc(var(--spacing)*4)}.py-0\\.5{padding-block:calc(var(--spacing)*.5)}.py-1{padding-block:calc(var(--spacing)*1)}.py-1\\.5{padding-block:calc(var(--spacing)*1.5)}.py-2{padding-block:calc(var(--spacing)*2)}.py-2\\.5{padding-block:calc(var(--spacing)*2.5)}.py-3{padding-block:calc(var(--spacing)*3)}.py-4{padding-block:calc(var(--spacing)*4)}.py-8{padding-block:calc(var(--spacing)*8)}.text-center{text-align:center}.text-left{text-align:left}.align-middle{vertical-align:middle}.align-top{vertical-align:top}.font-mono{font-family:var(--font-mono)}.text-lg{font-size:var(--text-lg);line-height:var(--tw-leading,var(--text-lg--line-height))}.text-sm{font-size:var(--text-sm);line-height:var(--tw-leading,var(--text-sm--line-height))}.text-xs{font-size:var(--text-xs);line-height:var(--tw-leading,var(--text-xs--line-height))}.text-\\[9px\\]{font-size:9px}.text-\\[10px\\]{font-size:10px}.text-\\[11px\\]{font-size:11px}.leading-relaxed{--tw-leading:var(--leading-relaxed);line-height:var(--leading-relaxed)}.leading-snug{--tw-leading:var(--leading-snug);line-height:var(--leading-snug)}.font-medium{--tw-font-weight:var(--font-weight-medium);font-weight:var(--font-weight-medium)}.font-semibold{--tw-font-weight:var(--font-weight-semibold);font-weight:var(--font-weight-semibold)}.break-words{overflow-wrap:break-word}.whitespace-nowrap{white-space:nowrap}.whitespace-pre-wrap{white-space:pre-wrap}.text-amber-600{color:var(--color-amber-600)}.text-amber-700{color:var(--color-amber-700)}.text-amber-800{color:var(--color-amber-800)}.text-amber-900{color:var(--color-amber-900)}.text-blue-600{color:var(--color-blue-600)}.text-blue-700{color:var(--color-blue-700)}.text-blue-800{color:var(--color-blue-800)}.text-emerald-700{color:var(--color-emerald-700)}.text-gray-300{color:var(--color-gray-300)}.text-gray-400{color:var(--color-gray-400)}.text-gray-500{color:var(--color-gray-500)}.text-gray-600{color:var(--color-gray-600)}.text-gray-700{color:var(--color-gray-700)}.text-gray-800{color:var(--color-gray-800)}.text-gray-900{color:var(--color-gray-900)}.text-green-700{color:var(--color-green-700)}.text-green-800{color:var(--color-green-800)}.text-purple-700{color:var(--color-purple-700)}.text-rose-700{color:var(--color-rose-700)}.text-rose-800{color:var(--color-rose-800)}.text-rose-900{color:var(--color-rose-900)}.text-sky-700{color:var(--color-sky-700)}.text-slate-800{color:var(--color-slate-800)}.text-slate-900{color:var(--color-slate-900)}.text-violet-700{color:var(--color-violet-700)}.text-white{color:var(--color-white)}.italic{font-style:italic}.opacity-0{opacity:0}.opacity-80{opacity:.8}.shadow-\\[0_12px_30px_rgba\\(15\\,23\\,42\\,0\\.12\\)\\]{--tw-shadow:0 12px 30px var(--tw-shadow-color,#0f172a1f);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.shadow-lg{--tw-shadow:0 10px 15px -3px var(--tw-shadow-color,#0000001a),0 4px 6px -4px var(--tw-shadow-color,#0000001a);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.shadow-md{--tw-shadow:0 4px 6px -1px var(--tw-shadow-color,#0000001a),0 2px 4px -2px var(--tw-shadow-color,#0000001a);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.shadow-sm{--tw-shadow:0 1px 3px 0 var(--tw-shadow-color,#0000001a),0 1px 2px -1px var(--tw-shadow-color,#0000001a);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.ring-1{--tw-ring-shadow:var(--tw-ring-inset,)0 0 0 calc(1px + var(--tw-ring-offset-width))var(--tw-ring-color,currentcolor);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.ring-black\\/5{--tw-ring-color:#0000000d}@supports (color:color-mix(in lab, red, red)){.ring-black\\/5{--tw-ring-color:color-mix(in oklab,var(--color-black)5%,transparent)}}.ring-transparent{--tw-ring-color:transparent}.ring-white\\/10{--tw-ring-color:#ffffff1a}@supports (color:color-mix(in lab, red, red)){.ring-white\\/10{--tw-ring-color:color-mix(in oklab,var(--color-white)10%,transparent)}}.backdrop-blur-md{--tw-backdrop-blur:blur(var(--blur-md));-webkit-backdrop-filter:var(--tw-backdrop-blur,)var(--tw-backdrop-brightness,)var(--tw-backdrop-contrast,)var(--tw-backdrop-grayscale,)var(--tw-backdrop-hue-rotate,)var(--tw-backdrop-invert,)var(--tw-backdrop-opacity,)var(--tw-backdrop-saturate,)var(--tw-backdrop-sepia,);backdrop-filter:var(--tw-backdrop-blur,)var(--tw-backdrop-brightness,)var(--tw-backdrop-contrast,)var(--tw-backdrop-grayscale,)var(--tw-backdrop-hue-rotate,)var(--tw-backdrop-invert,)var(--tw-backdrop-opacity,)var(--tw-backdrop-saturate,)var(--tw-backdrop-sepia,)}.backdrop-blur-sm{--tw-backdrop-blur:blur(var(--blur-sm));-webkit-backdrop-filter:var(--tw-backdrop-blur,)var(--tw-backdrop-brightness,)var(--tw-backdrop-contrast,)var(--tw-backdrop-grayscale,)var(--tw-backdrop-hue-rotate,)var(--tw-backdrop-invert,)var(--tw-backdrop-opacity,)var(--tw-backdrop-saturate,)var(--tw-backdrop-sepia,);backdrop-filter:var(--tw-backdrop-blur,)var(--tw-backdrop-brightness,)var(--tw-backdrop-contrast,)var(--tw-backdrop-grayscale,)var(--tw-backdrop-hue-rotate,)var(--tw-backdrop-invert,)var(--tw-backdrop-opacity,)var(--tw-backdrop-saturate,)var(--tw-backdrop-sepia,)}.transition{transition-property:color,background-color,border-color,outline-color,text-decoration-color,fill,stroke,--tw-gradient-from,--tw-gradient-via,--tw-gradient-to,opacity,box-shadow,transform,translate,scale,rotate,filter,-webkit-backdrop-filter,backdrop-filter,display,content-visibility,overlay,pointer-events;transition-timing-function:var(--tw-ease,var(--default-transition-timing-function));transition-duration:var(--tw-duration,var(--default-transition-duration))}.outline-none{--tw-outline-style:none;outline-style:none}.select-none{-webkit-user-select:none;user-select:none}@media (hover:hover){.group-hover\\:opacity-100:is(:where(.group):hover *){opacity:1}.hover\\:scale-105:hover{--tw-scale-x:105%;--tw-scale-y:105%;--tw-scale-z:105%;scale:var(--tw-scale-x)var(--tw-scale-y)}.hover\\:border-gray-300:hover{border-color:var(--color-gray-300)}.hover\\:border-white\\/30:hover{border-color:#ffffff4d}@supports (color:color-mix(in lab, red, red)){.hover\\:border-white\\/30:hover{border-color:color-mix(in oklab,var(--color-white)30%,transparent)}}.hover\\:bg-blue-50\\/50:hover{background-color:#eff6ff80}@supports (color:color-mix(in lab, red, red)){.hover\\:bg-blue-50\\/50:hover{background-color:color-mix(in oklab,var(--color-blue-50)50%,transparent)}}.hover\\:bg-gray-50:hover{background-color:var(--color-gray-50)}.hover\\:bg-gray-100:hover{background-color:var(--color-gray-100)}.hover\\:bg-gray-200:hover{background-color:var(--color-gray-200)}.hover\\:bg-gray-800:hover{background-color:var(--color-gray-800)}.hover\\:bg-slate-900\\/95:hover{background-color:#0f172bf2}@supports (color:color-mix(in lab, red, red)){.hover\\:bg-slate-900\\/95:hover{background-color:color-mix(in oklab,var(--color-slate-900)95%,transparent)}}.hover\\:text-gray-600:hover{color:var(--color-gray-600)}.hover\\:text-gray-900:hover{color:var(--color-gray-900)}}.focus\\:border-gray-300:focus{border-color:var(--color-gray-300)}.focus\\:bg-gray-50:focus{background-color:var(--color-gray-50)}.focus\\:ring-2:focus{--tw-ring-shadow:var(--tw-ring-inset,)0 0 0 calc(2px + var(--tw-ring-offset-width))var(--tw-ring-color,currentcolor);box-shadow:var(--tw-inset-shadow),var(--tw-inset-ring-shadow),var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow)}.focus\\:ring-gray-200:focus{--tw-ring-color:var(--color-gray-200)}.focus\\:outline-none:focus{--tw-outline-style:none;outline-style:none}.focus-visible\\:outline:focus-visible{outline-style:var(--tw-outline-style);outline-width:1px}.focus-visible\\:outline-2:focus-visible{outline-style:var(--tw-outline-style);outline-width:2px}.focus-visible\\:outline-offset-2:focus-visible{outline-offset:2px}.focus-visible\\:outline-gray-300:focus-visible{outline-color:var(--color-gray-300)}.focus-visible\\:outline-gray-400:focus-visible{outline-color:var(--color-gray-400)}.focus-visible\\:outline-rose-500:focus-visible{outline-color:var(--color-rose-500)}.disabled\\:cursor-not-allowed:disabled{cursor:not-allowed}.disabled\\:opacity-50:disabled{opacity:.5}@media (min-width:48rem){.md\\:grid-cols-4{grid-template-columns:repeat(4,minmax(0,1fr))}}.\\[\\&\\>svg\\]\\:\\!h-8>svg{height:calc(var(--spacing)*8)!important}.\\[\\&\\>svg\\]\\:\\!w-8>svg{width:calc(var(--spacing)*8)!important}.sr-only{clip:rect(0,0,0,0);white-space:nowrap;border-width:0;width:1px;height:1px;margin:-1px;padding:0;position:absolute;overflow:hidden}}@property --tw-translate-x{syntax:"*";inherits:false;initial-value:0}@property --tw-translate-y{syntax:"*";inherits:false;initial-value:0}@property --tw-translate-z{syntax:"*";inherits:false;initial-value:0}@property --tw-rotate-x{syntax:"*";inherits:false}@property --tw-rotate-y{syntax:"*";inherits:false}@property --tw-rotate-z{syntax:"*";inherits:false}@property --tw-skew-x{syntax:"*";inherits:false}@property --tw-skew-y{syntax:"*";inherits:false}@property --tw-space-y-reverse{syntax:"*";inherits:false;initial-value:0}@property --tw-divide-y-reverse{syntax:"*";inherits:false;initial-value:0}@property --tw-border-style{syntax:"*";inherits:false;initial-value:solid}@property --tw-leading{syntax:"*";inherits:false}@property --tw-font-weight{syntax:"*";inherits:false}@property --tw-shadow{syntax:"*";inherits:false;initial-value:0 0 #0000}@property --tw-shadow-color{syntax:"*";inherits:false}@property --tw-shadow-alpha{syntax:"<percentage>";inherits:false;initial-value:100%}@property --tw-inset-shadow{syntax:"*";inherits:false;initial-value:0 0 #0000}@property --tw-inset-shadow-color{syntax:"*";inherits:false}@property --tw-inset-shadow-alpha{syntax:"<percentage>";inherits:false;initial-value:100%}@property --tw-ring-color{syntax:"*";inherits:false}@property --tw-ring-shadow{syntax:"*";inherits:false;initial-value:0 0 #0000}@property --tw-inset-ring-color{syntax:"*";inherits:false}@property --tw-inset-ring-shadow{syntax:"*";inherits:false;initial-value:0 0 #0000}@property --tw-ring-inset{syntax:"*";inherits:false}@property --tw-ring-offset-width{syntax:"<length>";inherits:false;initial-value:0}@property --tw-ring-offset-color{syntax:"*";inherits:false;initial-value:#fff}@property --tw-ring-offset-shadow{syntax:"*";inherits:false;initial-value:0 0 #0000}@property --tw-backdrop-blur{syntax:"*";inherits:false}@property --tw-backdrop-brightness{syntax:"*";inherits:false}@property --tw-backdrop-contrast{syntax:"*";inherits:false}@property --tw-backdrop-grayscale{syntax:"*";inherits:false}@property --tw-backdrop-hue-rotate{syntax:"*";inherits:false}@property --tw-backdrop-invert{syntax:"*";inherits:false}@property --tw-backdrop-opacity{syntax:"*";inherits:false}@property --tw-backdrop-saturate{syntax:"*";inherits:false}@property --tw-backdrop-sepia{syntax:"*";inherits:false}@property --tw-scale-x{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-y{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-z{syntax:"*";inherits:false;initial-value:1}@property --tw-outline-style{syntax:"*";inherits:false;initial-value:solid}@keyframes pulse{50%{opacity:.5}}';

// src/assets/inspector-logo.svg
var inspector_logo_default = 'data:image/svg+xml,<svg width="136" height="26" viewBox="0 0 136 26" fill="none" xmlns="http://www.w3.org/2000/svg">%0A<path d="M15.5253 11.0828H11.2602C11.2299 10.7299 11.1492 10.4097 11.0181 10.1224C10.8921 9.83503 10.7157 9.588 10.4888 9.38129C10.267 9.16958 9.99725 9.00818 9.67964 8.89733C9.36202 8.78136 9.00155 8.72338 8.59823 8.72338C7.89242 8.72338 7.29501 8.89477 6.80598 9.23756C6.322 9.58045 5.95397 10.072 5.70189 10.7122C5.45486 11.3524 5.33134 12.1213 5.33134 13.0187C5.33134 13.9665 5.45738 14.7606 5.70946 15.4008C5.96657 16.0361 6.33712 16.515 6.82111 16.8376C7.30509 17.1553 7.88739 17.314 8.56798 17.314C8.95618 17.314 9.30404 17.2662 9.61158 17.1704C9.91911 17.0695 10.1863 16.9259 10.4132 16.7393C10.64 16.5528 10.8241 16.3284 10.9653 16.0663C11.1114 15.7991 11.2097 15.4991 11.2602 15.1664L15.5253 15.1967C15.4748 15.852 15.2908 16.52 14.9732 17.2006C14.6556 17.8762 14.2044 18.5013 13.6195 19.076C13.0398 19.6457 12.3214 20.1046 11.4643 20.4524C10.6073 20.8003 9.61158 20.9742 8.47724 20.9742C7.05554 20.9742 5.78004 20.6692 4.65074 20.0592C3.52649 19.4491 2.63667 18.5517 1.98126 17.367C1.33091 16.1823 1.00574 14.7328 1.00574 13.0187C1.00574 11.2946 1.33848 9.84259 2.00396 8.66285C2.66944 7.47811 3.56682 6.58324 4.69611 5.97827C5.82541 5.36825 7.08578 5.06323 8.47724 5.06323C9.45528 5.06323 10.3552 5.19683 11.177 5.46403C11.9987 5.73123 12.7196 6.12194 13.3397 6.63618C13.9599 7.14537 14.4589 7.77302 14.8371 8.51922C15.2152 9.26533 15.4446 10.1198 15.5253 11.0828Z" fill="%23010507"/>%0A<path d="M23.0894 20.9742C21.829 20.9742 20.7501 20.7246 19.8527 20.2256C18.9553 19.7214 18.2672 19.0206 17.7883 18.1232C17.3093 17.2208 17.0698 16.1747 17.0698 14.9849C17.0698 13.7951 17.3093 12.7516 17.7883 11.8542C18.2672 10.9517 18.9553 10.2509 19.8527 9.75184C20.7501 9.24767 21.829 8.99564 23.0894 8.99564C24.3498 8.99564 25.4286 9.24767 26.326 9.75184C27.2234 10.2509 27.9115 10.9517 28.3905 11.8542C28.8695 12.7516 29.1089 13.7951 29.1089 14.9849C29.1089 16.1747 28.8695 17.2208 28.3905 18.1232C27.9115 19.0206 27.2234 19.7214 26.326 20.2256C25.4286 20.7246 24.3498 20.9742 23.0894 20.9742ZM23.1196 17.8888C23.4726 17.8888 23.7776 17.7703 24.0346 17.5334C24.2918 17.2965 24.4909 16.9587 24.6321 16.5201C24.7733 16.0814 24.8438 15.5597 24.8438 14.9547C24.8438 14.3446 24.7733 13.8228 24.6321 13.3893C24.4909 12.9506 24.2918 12.6128 24.0346 12.3759C23.7776 12.139 23.4726 12.0205 23.1196 12.0205C22.7466 12.0205 22.4265 12.139 22.1592 12.3759C21.8921 12.6128 21.6878 12.9506 21.5466 13.3893C21.4056 13.8228 21.3349 14.3446 21.3349 14.9547C21.3349 15.5597 21.4056 16.0814 21.5466 16.5201C21.6878 16.9587 21.8921 17.2965 22.1592 17.5334C22.4265 17.7703 22.7466 17.8888 23.1196 17.8888Z" fill="%23010507"/>%0A<path d="M30.8029 25.1184V9.14681H34.947V11.1735H35.0377C35.189 10.7803 35.4109 10.4199 35.7032 10.0922C35.9957 9.7594 36.3587 9.4947 36.7922 9.2981C37.2257 9.09639 37.7299 8.99564 38.3046 8.99564C39.0709 8.99564 39.7995 9.19979 40.4901 9.60811C41.1858 10.0165 41.7505 10.6593 42.1841 11.5365C42.6227 12.4138 42.842 13.5531 42.842 14.9547C42.842 16.2957 42.6328 17.4073 42.2143 18.2896C41.801 19.1719 41.2464 19.8298 40.5506 20.2633C39.8599 20.697 39.1011 20.9137 38.2744 20.9137C37.7299 20.9137 37.2434 20.8255 36.8149 20.649C36.3914 20.4675 36.0284 20.223 35.7259 19.9155C35.4285 19.6029 35.199 19.25 35.0377 18.8568H34.9772V25.1184H30.8029ZM34.8865 14.9547C34.8865 15.5193 34.9596 16.0083 35.1058 16.4218C35.257 16.8301 35.4688 17.1477 35.7411 17.3746C36.0183 17.5964 36.3486 17.7073 36.7317 17.7073C37.1149 17.7073 37.44 17.5989 37.7072 17.3821C37.9795 17.1603 38.1862 16.8452 38.3273 16.4369C38.4736 16.0234 38.5466 15.5294 38.5466 14.9547C38.5466 14.3799 38.4736 13.8884 38.3273 13.48C38.1862 13.0666 37.9795 12.7516 37.7072 12.5347C37.44 12.3129 37.1149 12.202 36.7317 12.202C36.3486 12.202 36.0183 12.3129 35.7411 12.5347C35.4688 12.7516 35.257 13.0666 35.1058 13.48C34.9596 13.8884 34.8865 14.3799 34.8865 14.9547Z" fill="%23010507"/>%0A<path d="M44.5908 20.7625V9.14678H48.7652V20.7625H44.5908ZM46.6779 7.93683C46.1133 7.93683 45.6294 7.75033 45.2261 7.37726C44.8227 7.00419 44.621 6.55549 44.621 6.03118C44.621 5.50686 44.8227 5.05816 45.2261 4.68509C45.6294 4.31202 46.1133 4.12549 46.6779 4.12549C47.2477 4.12549 47.7316 4.31202 48.1299 4.68509C48.5332 5.05816 48.7349 5.50686 48.7349 6.03118C48.7349 6.55549 48.5332 7.00419 48.1299 7.37726C47.7316 7.75033 47.2477 7.93683 46.6779 7.93683Z" fill="%23010507"/>%0A<path d="M55.0664 5.27496V20.7625H50.892V5.27496H55.0664Z" fill="%23010507"/>%0A<path d="M62.7893 20.9742C61.5289 20.9742 60.4501 20.7246 59.5527 20.2256C58.6553 19.7214 57.9671 19.0206 57.4881 18.1232C57.0093 17.2208 56.7698 16.1747 56.7698 14.9849C56.7698 13.7951 57.0093 12.7516 57.4881 11.8542C57.9671 10.9517 58.6553 10.2509 59.5527 9.75184C60.4501 9.24767 61.5289 8.99564 62.7893 8.99564C64.0497 8.99564 65.1286 9.24767 66.026 9.75184C66.9234 10.2509 67.6115 10.9517 68.0904 11.8542C68.5694 12.7516 68.8089 13.7951 68.8089 14.9849C68.8089 16.1747 68.5694 17.2208 68.0904 18.1232C67.6115 19.0206 66.9234 19.7214 66.026 20.2256C65.1286 20.7246 64.0497 20.9742 62.7893 20.9742ZM62.8195 17.8888C63.1724 17.8888 63.4775 17.7703 63.7346 17.5334C63.9917 17.2965 64.1908 16.9587 64.332 16.5201C64.4731 16.0814 64.5438 15.5597 64.5438 14.9547C64.5438 14.3446 64.4731 13.8228 64.332 13.3893C64.1908 12.9506 63.9917 12.6128 63.7346 12.3759C63.4775 12.139 63.1724 12.0205 62.8195 12.0205C62.4464 12.0205 62.1263 12.139 61.8592 12.3759C61.5919 12.6128 61.3878 12.9506 61.2466 13.3893C61.1054 13.8228 61.0349 14.3446 61.0349 14.9547C61.0349 15.5597 61.1054 16.0814 61.2466 16.5201C61.3878 16.9587 61.5919 17.2965 61.8592 17.5334C62.1263 17.7703 62.4464 17.8888 62.8195 17.8888Z" fill="%23010507"/>%0A<path d="M77.4298 9.1468V12.1717H69.7769V9.1468H77.4298ZM71.2591 6.36392H75.4334V17.0267C75.4334 17.188 75.4612 17.3241 75.5166 17.4351C75.572 17.5409 75.6578 17.6216 75.7738 17.677C75.8897 17.7274 76.0384 17.7527 76.2199 17.7527C76.3459 17.7527 76.4921 17.7375 76.6585 17.7073C76.8299 17.677 76.9559 17.6518 77.0366 17.6317L77.6416 20.5659C77.4551 20.6213 77.1879 20.6894 76.84 20.77C76.4972 20.8507 76.0888 20.9037 75.6149 20.9288C74.657 20.9792 73.8529 20.8809 73.2026 20.6339C72.5522 20.3818 72.0631 19.9861 71.7355 19.4466C71.4078 18.9072 71.249 18.2316 71.2591 17.42V6.36392Z" fill="%23010507"/>%0A<path d="M79.3223 20.7625V5.27496H83.5268V11.5365H83.7385L88.397 5.27496H93.2973L88.0642 12.1717L93.4183 20.7625H88.397L84.9183 14.9546L83.5268 16.7696V20.7625H79.3223Z" fill="%23010507"/>%0A<path d="M95.3334 20.7625V9.14678H99.5077V20.7625H95.3334ZM97.4206 7.93683C96.8559 7.93683 96.3719 7.75033 95.9686 7.37726C95.5653 7.00419 95.3637 6.55549 95.3637 6.03118C95.3637 5.50686 95.5653 5.05816 95.9686 4.68509C96.3719 4.31202 96.8559 4.12549 97.4206 4.12549C97.9902 4.12549 98.4743 4.31202 98.8725 4.68509C99.2758 5.05816 99.4775 5.50686 99.4775 6.03118C99.4775 6.55549 99.2758 7.00419 98.8725 7.37726C98.4743 7.75033 97.9902 7.93683 97.4206 7.93683Z" fill="%23010507"/>%0A<path d="M108.562 9.1468V12.1717H100.909V9.1468H108.562ZM102.391 6.36392H106.565V17.0267C106.565 17.188 106.593 17.3241 106.648 17.4351C106.704 17.5409 106.789 17.6216 106.905 17.677C107.021 17.7274 107.17 17.7527 107.352 17.7527C107.478 17.7527 107.623 17.7375 107.791 17.7073C107.962 17.677 108.087 17.6518 108.168 17.6317L108.774 20.5659C108.587 20.6213 108.319 20.6894 107.971 20.77C107.629 20.8507 107.22 20.9037 106.747 20.9288C105.789 20.9792 104.985 20.8809 104.334 20.6339C103.684 20.3818 103.195 19.9861 102.867 19.4466C102.54 18.9072 102.381 18.2316 102.391 17.42V6.36392Z" fill="%23010507"/>%0A<path d="M119.539 8.67678C121.648 5.91866 123.398 3.19128 124.071 0.988777C124.089 0.928849 124.159 0.90339 124.212 0.937987C126.553 2.48864 130.818 3.50932 134.591 3.53328C134.656 3.53369 134.7 3.59769 134.677 3.65822C133.423 6.84039 131.891 12.5423 131.831 19.0536C131.831 19.1503 131.695 19.185 131.647 19.1009C129.5 15.3438 122.623 10.0644 119.574 8.81838C119.518 8.79519 119.502 8.7255 119.539 8.67678Z" fill="url(%23paint0_linear_8010_10612)"/>%0A<path d="M126.653 6.98969C123.357 8.03321 120.345 8.61334 119.626 8.74516C119.58 8.75356 119.571 8.81686 119.614 8.83473C122.687 10.1121 129.53 15.3761 131.657 19.118C131.661 19.1262 131.671 19.1292 131.68 19.1255C131.688 19.1214 131.693 19.1108 131.69 19.1016L126.653 6.98969Z" fill="url(%23paint1_linear_8010_10612)"/>%0A<path d="M124.221 0.931179C127.043 2.4702 130.303 3.16142 134.629 3.52564C134.656 3.52796 134.665 3.56437 134.641 3.57702C134.088 3.86136 130.918 5.47409 128.565 6.33784C127.934 6.56926 127.3 6.78394 126.676 6.982C126.662 6.98633 126.647 6.97951 126.641 6.96631L124.156 0.989469C124.139 0.949222 124.183 0.910305 124.221 0.931179Z" fill="url(%23paint2_linear_8010_10612)"/>%0A<path d="M124.171 1.02203L131.742 19.082" stroke="%23513C9F" stroke-width="0.183881" stroke-linecap="round"/>%0A<path d="M119.628 8.74279C119.628 8.74279 123.809 7.99233 127.739 6.63677C131.668 5.28124 134.594 3.6214 134.594 3.6214" stroke="%23513C9F" stroke-width="0.183881" stroke-linecap="round"/>%0A<path d="M125.209 3.30383L122.405 12.6358M122.405 12.6358H129.07M122.405 12.6358L111.874 25.0383" stroke="%23ABABAB" stroke-width="0.321797" stroke-linecap="round"/>%0A<path d="M119.181 22.485L117.94 22.6596C118.584 24.3618 119.904 25.1053 121.479 25.1053C125.341 25.1053 124.163 20.7382 126.4 20.7382C128.023 20.7382 127.364 24.2778 130.857 24.2778C132.989 24.2778 133.201 22.1301 132.837 21.2061C132.835 21.2005 132.833 21.1953 132.83 21.1902L132.259 20.3154C132.222 20.2572 132.131 20.2791 132.125 20.3483L132.018 21.4087C132.011 21.4824 132.013 21.5559 132.021 21.6295C132.109 22.3621 132.165 24.14 130.857 24.14C129.477 24.14 129.145 20.6462 126.4 20.6462C123.181 20.6462 123.594 24.9675 121.618 24.9675C120.313 24.9675 119.319 23.4964 119.181 22.485Z" fill="url(%23paint3_linear_8010_10612)"/>%0A<defs>%0A<linearGradient id="paint0_linear_8010_10612" x1="129.301" y1="2.33848" x2="125.623" y2="12.452" gradientUnits="userSpaceOnUse">%0A<stop stop-color="%236430AB"/>%0A<stop offset="1" stop-color="%23AA89D8"/>%0A</linearGradient>%0A<linearGradient id="paint1_linear_8010_10612" x1="126.451" y1="8.03874" x2="121.717" y2="17.1869" gradientUnits="userSpaceOnUse">%0A<stop stop-color="%23005DBB"/>%0A<stop offset="1" stop-color="%233D92E8"/>%0A</linearGradient>%0A<linearGradient id="paint2_linear_8010_10612" x1="128.565" y1="2.33842" x2="127.139" y2="6.79755" gradientUnits="userSpaceOnUse">%0A<stop stop-color="%231B70C4"/>%0A<stop offset="1" stop-color="%2354A4F2"/>%0A</linearGradient>%0A<linearGradient id="paint3_linear_8010_10612" x1="117.94" y1="22.7838" x2="132.981" y2="22.7838" gradientUnits="userSpaceOnUse">%0A<stop stop-color="%234497EA"/>%0A<stop offset="0.254755" stop-color="%231463B2"/>%0A<stop offset="0.498725" stop-color="%230A437D"/>%0A<stop offset="0.666667" stop-color="%232476C8"/>%0A<stop offset="0.972542" stop-color="%230C549A"/>%0A</linearGradient>%0A</defs>%0A</svg>%0A';

// src/assets/inspector-logo-icon.svg
var inspector_logo_icon_default = 'data:image/svg+xml,<?xml version="1.0" encoding="utf-8"?>%0A<svg xmlns="http://www.w3.org/2000/svg" width="55" height="60" viewBox="0 0 55 60" fill="none">%0A  <g stroke="%23FFFFFF" stroke-width="3.3125" stroke-linejoin="round">%0A    <path d="M0 21.9336L16.5449 0C40.8483 5.37332 53 8.05998 53 8.05998L43.123 56L0 21.9336Z" />%0A    <line x1="16.5828" y1="0" x2="43.2454" y2="56" />%0A    <line x1="0" y1="21.9336" x2="53" y2="8.48421" />%0A  </g>%0A</svg>%0A';

// src/index.ts
import { unsafeHTML } from "lit/directives/unsafe-html.js";
import { marked } from "marked";
import { icons } from "lucide";
import {
  CopilotKitCoreRuntimeConnectionStatus
} from "@copilotkitnext/core";

// src/lib/context-helpers.ts
function updateSizeFromElement(state, element, fallback) {
  const rect = element.getBoundingClientRect();
  state.size = {
    width: rect.width || fallback.width,
    height: rect.height || fallback.height
  };
}
function clampSize(size, viewport, edgeMargin, minWidth, minHeight) {
  const maxWidth = Math.max(minWidth, viewport.width - edgeMargin * 2);
  const maxHeight = Math.max(minHeight, viewport.height - edgeMargin * 2);
  return {
    width: clamp(size.width, minWidth, maxWidth),
    height: clamp(size.height, minHeight, maxHeight)
  };
}
function constrainToViewport(state, position, viewport, edgeMargin) {
  const maxX = Math.max(edgeMargin, viewport.width - state.size.width - edgeMargin);
  const maxY = Math.max(edgeMargin, viewport.height - state.size.height - edgeMargin);
  return {
    x: clamp(position.x, edgeMargin, maxX),
    y: clamp(position.y, edgeMargin, maxY)
  };
}
function keepPositionWithinViewport(state, viewport, edgeMargin) {
  state.position = constrainToViewport(state, state.position, viewport, edgeMargin);
}
function centerContext(state, viewport, edgeMargin) {
  const centered = {
    x: Math.round((viewport.width - state.size.width) / 2),
    y: Math.round((viewport.height - state.size.height) / 2)
  };
  state.position = constrainToViewport(state, centered, viewport, edgeMargin);
  updateAnchorFromPosition(state, viewport, edgeMargin);
  return state.position;
}
function updateAnchorFromPosition(state, viewport, edgeMargin) {
  const centerX = state.position.x + state.size.width / 2;
  const centerY = state.position.y + state.size.height / 2;
  const horizontal = centerX < viewport.width / 2 ? "left" : "right";
  const vertical = centerY < viewport.height / 2 ? "top" : "bottom";
  state.anchor = { horizontal, vertical };
  const maxHorizontalOffset = Math.max(edgeMargin, viewport.width - state.size.width - edgeMargin);
  const maxVerticalOffset = Math.max(edgeMargin, viewport.height - state.size.height - edgeMargin);
  state.anchorOffset = {
    x: horizontal === "left" ? clamp(state.position.x, edgeMargin, maxHorizontalOffset) : clamp(viewport.width - state.position.x - state.size.width, edgeMargin, maxHorizontalOffset),
    y: vertical === "top" ? clamp(state.position.y, edgeMargin, maxVerticalOffset) : clamp(viewport.height - state.position.y - state.size.height, edgeMargin, maxVerticalOffset)
  };
}
function applyAnchorPosition(state, viewport, edgeMargin) {
  const maxHorizontalOffset = Math.max(edgeMargin, viewport.width - state.size.width - edgeMargin);
  const maxVerticalOffset = Math.max(edgeMargin, viewport.height - state.size.height - edgeMargin);
  const horizontalOffset = clamp(state.anchorOffset.x, edgeMargin, maxHorizontalOffset);
  const verticalOffset = clamp(state.anchorOffset.y, edgeMargin, maxVerticalOffset);
  const x = state.anchor.horizontal === "left" ? horizontalOffset : viewport.width - state.size.width - horizontalOffset;
  const y = state.anchor.vertical === "top" ? verticalOffset : viewport.height - state.size.height - verticalOffset;
  state.anchorOffset = { x: horizontalOffset, y: verticalOffset };
  state.position = constrainToViewport(state, { x, y }, viewport, edgeMargin);
  return state.position;
}
function clamp(value, min, max) {
  return Math.min(Math.max(min, value), max);
}

// src/lib/persistence.ts
function loadInspectorState(storageKey) {
  if (typeof window === "undefined") {
    return null;
  }
  const raw = window.localStorage.getItem(storageKey);
  if (raw) {
    try {
      const parsed = JSON.parse(raw);
      if (parsed && typeof parsed === "object") {
        return parsed;
      }
    } catch {
    }
  }
  if (typeof document !== "undefined") {
    const prefix = `${storageKey}=`;
    const entry = document.cookie.split("; ").find((cookie) => cookie.startsWith(prefix));
    if (entry) {
      const legacyRaw = entry.substring(prefix.length);
      try {
        const parsed = JSON.parse(decodeURIComponent(legacyRaw));
        if (parsed && typeof parsed === "object") {
          return parsed;
        }
      } catch {
        return null;
      }
    }
  }
  return null;
}
function saveInspectorState(storageKey, state) {
  if (typeof window === "undefined") {
    return;
  }
  try {
    window.localStorage.setItem(storageKey, JSON.stringify(state));
  } catch (error) {
    console.warn("Failed to persist inspector state", error);
  }
}
function isValidAnchor(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value;
  return (candidate.horizontal === "left" || candidate.horizontal === "right") && (candidate.vertical === "top" || candidate.vertical === "bottom");
}
function isValidPosition(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value;
  return isFiniteNumber(candidate.x) && isFiniteNumber(candidate.y);
}
function isValidSize(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value;
  return isFiniteNumber(candidate.width) && isFiniteNumber(candidate.height);
}
function isFiniteNumber(value) {
  return typeof value === "number" && Number.isFinite(value);
}
function isValidDockMode(value) {
  return value === "floating" || value === "docked-left";
}

// src/index.ts
var WEB_INSPECTOR_TAG = "cpk-web-inspector";
var EDGE_MARGIN = 16;
var DRAG_THRESHOLD = 6;
var MIN_WINDOW_WIDTH = 600;
var MIN_WINDOW_WIDTH_DOCKED_LEFT = 420;
var MIN_WINDOW_HEIGHT = 200;
var INSPECTOR_STORAGE_KEY = "cpk:inspector:state";
var ANNOUNCEMENT_STORAGE_KEY = "cpk:inspector:announcements";
var ANNOUNCEMENT_URL = "https://cdn.copilotkit.ai/announcements.json";
var DEFAULT_BUTTON_SIZE = { width: 48, height: 48 };
var DEFAULT_WINDOW_SIZE = { width: 840, height: 560 };
var DOCKED_LEFT_WIDTH = 500;
var MAX_AGENT_EVENTS = 200;
var MAX_TOTAL_EVENTS = 500;
var AGENT_EVENT_TYPES = [
  "RUN_STARTED",
  "RUN_FINISHED",
  "RUN_ERROR",
  "TEXT_MESSAGE_START",
  "TEXT_MESSAGE_CONTENT",
  "TEXT_MESSAGE_END",
  "TOOL_CALL_START",
  "TOOL_CALL_ARGS",
  "TOOL_CALL_END",
  "TOOL_CALL_RESULT",
  "STATE_SNAPSHOT",
  "STATE_DELTA",
  "MESSAGES_SNAPSHOT",
  "RAW_EVENT",
  "CUSTOM_EVENT"
];
var WebInspectorElement = class extends LitElement {
  constructor() {
    super(...arguments);
    this._core = null;
    this.coreSubscriber = null;
    this.coreUnsubscribe = null;
    this.runtimeStatus = null;
    this.coreProperties = {};
    this.lastCoreError = null;
    this.agentSubscriptions = /* @__PURE__ */ new Map();
    this.agentEvents = /* @__PURE__ */ new Map();
    this.agentMessages = /* @__PURE__ */ new Map();
    this.agentStates = /* @__PURE__ */ new Map();
    this.flattenedEvents = [];
    this.eventCounter = 0;
    this.contextStore = {};
    this.pointerId = null;
    this.dragStart = null;
    this.dragOffset = { x: 0, y: 0 };
    this.isDragging = false;
    this.pointerContext = null;
    this.isOpen = false;
    this.draggedDuringInteraction = false;
    this.ignoreNextButtonClick = false;
    this.selectedMenu = "ag-ui-events";
    this.contextMenuOpen = false;
    this.dockMode = "floating";
    this.previousBodyMargins = null;
    this.transitionTimeoutId = null;
    this.pendingSelectedContext = null;
    this.autoAttachCore = true;
    this.attemptedAutoAttach = false;
    this.cachedTools = [];
    this.toolSignature = "";
    this.eventFilterText = "";
    this.eventTypeFilter = "all";
    this.announcementMarkdown = null;
    this.announcementHtml = null;
    this.announcementTimestamp = null;
    this.announcementPreviewText = null;
    this.hasUnseenAnnouncement = false;
    this.announcementLoaded = false;
    this.announcementLoadError = null;
    this.announcementPromise = null;
    this.showAnnouncementPreview = true;
    this.contextState = {
      button: {
        position: { x: EDGE_MARGIN, y: EDGE_MARGIN },
        size: { ...DEFAULT_BUTTON_SIZE },
        anchor: { horizontal: "right", vertical: "top" },
        anchorOffset: { x: EDGE_MARGIN, y: EDGE_MARGIN }
      },
      window: {
        position: { x: EDGE_MARGIN, y: EDGE_MARGIN },
        size: { ...DEFAULT_WINDOW_SIZE },
        anchor: { horizontal: "right", vertical: "top" },
        anchorOffset: { x: EDGE_MARGIN, y: EDGE_MARGIN }
      }
    };
    this.hasCustomPosition = {
      button: false,
      window: false
    };
    this.resizePointerId = null;
    this.resizeStart = null;
    this.resizeInitialSize = null;
    this.isResizing = false;
    this.menuItems = [
      { key: "ag-ui-events", label: "AG-UI Events", icon: "Zap" },
      { key: "agents", label: "Agent", icon: "Bot" },
      { key: "frontend-tools", label: "Frontend Tools", icon: "Hammer" },
      { key: "agent-context", label: "Context", icon: "FileText" }
    ];
    this.handlePointerDown = (event) => {
      if (this.dockMode !== "floating" && this.isOpen) {
        return;
      }
      const target = event.currentTarget;
      const contextAttr = target?.dataset.dragContext;
      const context = contextAttr === "window" ? "window" : "button";
      const eventTarget = event.target;
      if (context === "window" && eventTarget?.closest("button")) {
        return;
      }
      this.pointerContext = context;
      this.measureContext(context);
      event.preventDefault();
      this.pointerId = event.pointerId;
      this.dragStart = { x: event.clientX, y: event.clientY };
      const state = this.contextState[context];
      this.dragOffset = {
        x: event.clientX - state.position.x,
        y: event.clientY - state.position.y
      };
      this.isDragging = false;
      this.draggedDuringInteraction = false;
      this.ignoreNextButtonClick = false;
      target?.setPointerCapture?.(this.pointerId);
    };
    this.handlePointerMove = (event) => {
      if (this.pointerId !== event.pointerId || !this.dragStart || !this.pointerContext) {
        return;
      }
      const distance = Math.hypot(event.clientX - this.dragStart.x, event.clientY - this.dragStart.y);
      if (!this.isDragging && distance < DRAG_THRESHOLD) {
        return;
      }
      event.preventDefault();
      this.setDragging(true);
      this.draggedDuringInteraction = true;
      const desired = {
        x: event.clientX - this.dragOffset.x,
        y: event.clientY - this.dragOffset.y
      };
      const constrained = this.constrainToViewport(desired, this.pointerContext);
      this.contextState[this.pointerContext].position = constrained;
      this.updateHostTransform(this.pointerContext);
    };
    this.handlePointerUp = (event) => {
      if (this.pointerId !== event.pointerId) {
        return;
      }
      const target = event.currentTarget;
      if (target?.hasPointerCapture(this.pointerId)) {
        target.releasePointerCapture(this.pointerId);
      }
      const context = this.pointerContext ?? this.activeContext;
      if (this.isDragging && this.pointerContext) {
        event.preventDefault();
        this.setDragging(false);
        if (this.pointerContext === "window") {
          this.updateAnchorFromPosition(this.pointerContext);
          this.hasCustomPosition.window = true;
          this.applyAnchorPosition(this.pointerContext);
        } else if (this.pointerContext === "button") {
          this.snapButtonToCorner();
          this.hasCustomPosition.button = true;
          if (this.draggedDuringInteraction) {
            this.ignoreNextButtonClick = true;
          }
        }
      } else if (context === "button" && !this.isOpen && !this.draggedDuringInteraction) {
        this.openInspector();
      }
      this.resetPointerTracking();
    };
    this.handlePointerCancel = (event) => {
      if (this.pointerId !== event.pointerId) {
        return;
      }
      const target = event.currentTarget;
      if (target?.hasPointerCapture(this.pointerId)) {
        target.releasePointerCapture(this.pointerId);
      }
      this.resetPointerTracking();
    };
    this.handleButtonClick = (event) => {
      if (this.isDragging) {
        event.preventDefault();
        return;
      }
      if (this.ignoreNextButtonClick) {
        event.preventDefault();
        this.ignoreNextButtonClick = false;
        return;
      }
      if (!this.isOpen) {
        event.preventDefault();
        this.openInspector();
      }
    };
    this.handleClosePointerDown = (event) => {
      event.stopPropagation();
      event.preventDefault();
    };
    this.handleCloseClick = () => {
      this.closeInspector();
    };
    this.handleResizePointerDown = (event) => {
      event.stopPropagation();
      event.preventDefault();
      this.hasCustomPosition.window = true;
      this.isResizing = true;
      this.resizePointerId = event.pointerId;
      this.resizeStart = { x: event.clientX, y: event.clientY };
      this.resizeInitialSize = { ...this.contextState.window.size };
      if (document.body && this.dockMode !== "floating") {
        document.body.style.transition = "";
      }
      const target = event.currentTarget;
      target?.setPointerCapture?.(event.pointerId);
    };
    this.handleResizePointerMove = (event) => {
      if (!this.isResizing || this.resizePointerId !== event.pointerId || !this.resizeStart || !this.resizeInitialSize) {
        return;
      }
      event.preventDefault();
      const deltaX = event.clientX - this.resizeStart.x;
      const deltaY = event.clientY - this.resizeStart.y;
      const state = this.contextState.window;
      if (this.dockMode === "docked-left") {
        state.size = this.clampWindowSize({
          width: this.resizeInitialSize.width + deltaX,
          height: state.size.height
        });
        if (document.body) {
          document.body.style.marginLeft = `${state.size.width}px`;
        }
      } else {
        state.size = this.clampWindowSize({
          width: this.resizeInitialSize.width + deltaX,
          height: this.resizeInitialSize.height + deltaY
        });
        this.keepPositionWithinViewport("window");
        this.updateAnchorFromPosition("window");
      }
      this.requestUpdate();
      this.updateHostTransform("window");
    };
    this.handleResizePointerUp = (event) => {
      if (this.resizePointerId !== event.pointerId) {
        return;
      }
      const target = event.currentTarget;
      if (target?.hasPointerCapture(this.resizePointerId)) {
        target.releasePointerCapture(this.resizePointerId);
      }
      if (this.dockMode === "floating") {
        this.updateAnchorFromPosition("window");
        this.applyAnchorPosition("window");
      }
      this.persistState();
      this.resetResizeTracking();
    };
    this.handleResizePointerCancel = (event) => {
      if (this.resizePointerId !== event.pointerId) {
        return;
      }
      const target = event.currentTarget;
      if (target?.hasPointerCapture(this.resizePointerId)) {
        target.releasePointerCapture(this.resizePointerId);
      }
      if (this.dockMode === "floating") {
        this.updateAnchorFromPosition("window");
        this.applyAnchorPosition("window");
      }
      this.persistState();
      this.resetResizeTracking();
    };
    this.handleResize = () => {
      this.measureContext("button");
      this.applyAnchorPosition("button");
      this.measureContext("window");
      if (this.hasCustomPosition.window) {
        this.applyAnchorPosition("window");
      } else {
        this.centerContext("window");
      }
      this.updateHostTransform();
    };
    this.contextOptions = [
      { key: "all-agents", label: "All Agents" }
    ];
    this.selectedContext = "all-agents";
    this.expandedRows = /* @__PURE__ */ new Set();
    this.copiedEvents = /* @__PURE__ */ new Set();
    this.expandedTools = /* @__PURE__ */ new Set();
    this.expandedContextItems = /* @__PURE__ */ new Set();
    this.copiedContextItems = /* @__PURE__ */ new Set();
    this.handleClearEvents = () => {
      if (this.selectedContext === "all-agents") {
        this.agentEvents.clear();
        this.flattenedEvents = [];
      } else {
        this.agentEvents.delete(this.selectedContext);
        this.flattenedEvents = this.flattenedEvents.filter((event) => event.agentId !== this.selectedContext);
      }
      this.expandedRows.clear();
      this.copiedEvents.clear();
      this.requestUpdate();
    };
    this.handleGlobalPointerDown = (event) => {
      if (!this.contextMenuOpen) {
        return;
      }
      const clickedDropdown = event.composedPath().some((node) => {
        return node instanceof HTMLElement && node.dataset?.contextDropdownRoot === "true";
      });
      if (!clickedDropdown) {
        this.contextMenuOpen = false;
        this.requestUpdate();
      }
    };
    this.handleDismissAnnouncement = () => {
      this.markAnnouncementSeen();
    };
  }
  static {
    this.properties = {
      core: { attribute: false },
      autoAttachCore: { type: Boolean, attribute: "auto-attach-core" }
    };
  }
  get core() {
    return this._core;
  }
  set core(value) {
    const oldValue = this._core;
    if (oldValue === value) {
      return;
    }
    this.detachFromCore();
    this._core = value ?? null;
    this.requestUpdate("core", oldValue);
    if (this._core) {
      this.attachToCore(this._core);
    }
  }
  attachToCore(core) {
    this.runtimeStatus = core.runtimeConnectionStatus;
    this.coreProperties = core.properties;
    this.lastCoreError = null;
    this.coreSubscriber = {
      onRuntimeConnectionStatusChanged: ({ status }) => {
        this.runtimeStatus = status;
        this.requestUpdate();
      },
      onPropertiesChanged: ({ properties }) => {
        this.coreProperties = properties;
        this.requestUpdate();
      },
      onError: ({ code, error }) => {
        this.lastCoreError = { code, message: error.message };
        this.requestUpdate();
      },
      onAgentsChanged: ({ agents }) => {
        this.processAgentsChanged(agents);
      },
      onContextChanged: ({ context }) => {
        this.contextStore = this.normalizeContextStore(context);
        this.requestUpdate();
      }
    };
    this.coreUnsubscribe = core.subscribe(this.coreSubscriber).unsubscribe;
    this.processAgentsChanged(core.agents);
    if (core.context) {
      this.contextStore = this.normalizeContextStore(core.context);
    }
  }
  detachFromCore() {
    if (this.coreUnsubscribe) {
      this.coreUnsubscribe();
      this.coreUnsubscribe = null;
    }
    this.coreSubscriber = null;
    this.runtimeStatus = null;
    this.lastCoreError = null;
    this.coreProperties = {};
    this.cachedTools = [];
    this.toolSignature = "";
    this.teardownAgentSubscriptions();
  }
  teardownAgentSubscriptions() {
    for (const unsubscribe of this.agentSubscriptions.values()) {
      unsubscribe();
    }
    this.agentSubscriptions.clear();
    this.agentEvents.clear();
    this.agentMessages.clear();
    this.agentStates.clear();
    this.flattenedEvents = [];
    this.eventCounter = 0;
  }
  processAgentsChanged(agents) {
    const seenAgentIds = /* @__PURE__ */ new Set();
    for (const agent of Object.values(agents)) {
      if (!agent?.agentId) {
        continue;
      }
      seenAgentIds.add(agent.agentId);
      this.subscribeToAgent(agent);
    }
    for (const agentId of Array.from(this.agentSubscriptions.keys())) {
      if (!seenAgentIds.has(agentId)) {
        this.unsubscribeFromAgent(agentId);
        this.agentEvents.delete(agentId);
        this.agentMessages.delete(agentId);
        this.agentStates.delete(agentId);
      }
    }
    this.updateContextOptions(seenAgentIds);
    this.refreshToolsSnapshot();
    this.requestUpdate();
  }
  refreshToolsSnapshot() {
    if (!this._core) {
      if (this.cachedTools.length > 0) {
        this.cachedTools = [];
        this.toolSignature = "";
        this.requestUpdate();
      }
      return;
    }
    const tools = this.extractToolsFromAgents();
    const signature = JSON.stringify(
      tools.map((tool) => ({
        agentId: tool.agentId,
        name: tool.name,
        type: tool.type,
        hasDescription: Boolean(tool.description),
        hasParameters: Boolean(tool.parameters)
      }))
    );
    if (signature !== this.toolSignature) {
      this.toolSignature = signature;
      this.cachedTools = tools;
      this.requestUpdate();
    }
  }
  tryAutoAttachCore() {
    if (this.attemptedAutoAttach || this._core || !this.autoAttachCore || typeof window === "undefined") {
      return;
    }
    this.attemptedAutoAttach = true;
    const globalWindow = window;
    const globalCandidates = [
      // Common app-level globals used during development
      globalWindow.__COPILOTKIT_CORE__,
      globalWindow.copilotkit?.core,
      globalWindow.copilotkitCore
    ];
    const foundCore = globalCandidates.find(
      (candidate) => !!candidate && typeof candidate === "object"
    );
    if (foundCore) {
      this.core = foundCore;
    }
  }
  subscribeToAgent(agent) {
    if (!agent.agentId) {
      return;
    }
    const agentId = agent.agentId;
    this.unsubscribeFromAgent(agentId);
    const subscriber = {
      onRunStartedEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "RUN_STARTED", event);
      },
      onRunFinishedEvent: ({ event, result }) => {
        this.recordAgentEvent(agentId, "RUN_FINISHED", { event, result });
      },
      onRunErrorEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "RUN_ERROR", event);
      },
      onTextMessageStartEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "TEXT_MESSAGE_START", event);
      },
      onTextMessageContentEvent: ({ event, textMessageBuffer }) => {
        this.recordAgentEvent(agentId, "TEXT_MESSAGE_CONTENT", { event, textMessageBuffer });
      },
      onTextMessageEndEvent: ({ event, textMessageBuffer }) => {
        this.recordAgentEvent(agentId, "TEXT_MESSAGE_END", { event, textMessageBuffer });
      },
      onToolCallStartEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "TOOL_CALL_START", event);
      },
      onToolCallArgsEvent: ({ event, toolCallBuffer, toolCallName, partialToolCallArgs }) => {
        this.recordAgentEvent(agentId, "TOOL_CALL_ARGS", { event, toolCallBuffer, toolCallName, partialToolCallArgs });
      },
      onToolCallEndEvent: ({ event, toolCallArgs, toolCallName }) => {
        this.recordAgentEvent(agentId, "TOOL_CALL_END", { event, toolCallArgs, toolCallName });
      },
      onToolCallResultEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "TOOL_CALL_RESULT", event);
      },
      onStateSnapshotEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "STATE_SNAPSHOT", event);
        this.syncAgentState(agent);
      },
      onStateDeltaEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "STATE_DELTA", event);
        this.syncAgentState(agent);
      },
      onMessagesSnapshotEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "MESSAGES_SNAPSHOT", event);
        this.syncAgentMessages(agent);
      },
      onMessagesChanged: () => {
        this.syncAgentMessages(agent);
      },
      onRawEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "RAW_EVENT", event);
      },
      onCustomEvent: ({ event }) => {
        this.recordAgentEvent(agentId, "CUSTOM_EVENT", event);
      }
    };
    const { unsubscribe } = agent.subscribe(subscriber);
    this.agentSubscriptions.set(agentId, unsubscribe);
    this.syncAgentMessages(agent);
    this.syncAgentState(agent);
    if (!this.agentEvents.has(agentId)) {
      this.agentEvents.set(agentId, []);
    }
  }
  unsubscribeFromAgent(agentId) {
    const unsubscribe = this.agentSubscriptions.get(agentId);
    if (unsubscribe) {
      unsubscribe();
      this.agentSubscriptions.delete(agentId);
    }
  }
  recordAgentEvent(agentId, type, payload) {
    const eventId = `${agentId}:${++this.eventCounter}`;
    const normalizedPayload = this.normalizeEventPayload(type, payload);
    const event = {
      id: eventId,
      agentId,
      type,
      timestamp: Date.now(),
      payload: normalizedPayload
    };
    const currentAgentEvents = this.agentEvents.get(agentId) ?? [];
    const nextAgentEvents = [event, ...currentAgentEvents].slice(0, MAX_AGENT_EVENTS);
    this.agentEvents.set(agentId, nextAgentEvents);
    this.flattenedEvents = [event, ...this.flattenedEvents].slice(0, MAX_TOTAL_EVENTS);
    this.refreshToolsSnapshot();
    this.requestUpdate();
  }
  syncAgentMessages(agent) {
    if (!agent?.agentId) {
      return;
    }
    const messages = this.normalizeAgentMessages(agent.messages);
    if (messages) {
      this.agentMessages.set(agent.agentId, messages);
    } else {
      this.agentMessages.delete(agent.agentId);
    }
    this.requestUpdate();
  }
  syncAgentState(agent) {
    if (!agent?.agentId) {
      return;
    }
    const state = agent.state;
    if (state === void 0 || state === null) {
      this.agentStates.delete(agent.agentId);
    } else {
      this.agentStates.set(agent.agentId, this.sanitizeForLogging(state));
    }
    this.requestUpdate();
  }
  updateContextOptions(agentIds) {
    const nextOptions = [
      { key: "all-agents", label: "All Agents" },
      ...Array.from(agentIds).sort((a, b) => a.localeCompare(b)).map((id) => ({ key: id, label: id }))
    ];
    const optionsChanged = this.contextOptions.length !== nextOptions.length || this.contextOptions.some((option, index) => option.key !== nextOptions[index]?.key);
    if (optionsChanged) {
      this.contextOptions = nextOptions;
    }
    const pendingContext = this.pendingSelectedContext;
    if (pendingContext) {
      const isPendingAvailable = pendingContext === "all-agents" || agentIds.has(pendingContext);
      if (isPendingAvailable) {
        if (this.selectedContext !== pendingContext) {
          this.selectedContext = pendingContext;
          this.expandedRows.clear();
        }
        this.pendingSelectedContext = null;
      } else if (agentIds.size > 0) {
        this.pendingSelectedContext = null;
      }
    }
    const hasSelectedContext = nextOptions.some((option) => option.key === this.selectedContext);
    if (!hasSelectedContext && this.pendingSelectedContext === null) {
      let nextSelected = "all-agents";
      if (agentIds.has("default")) {
        nextSelected = "default";
      } else if (agentIds.size > 0) {
        nextSelected = Array.from(agentIds).sort((a, b) => a.localeCompare(b))[0];
      }
      if (this.selectedContext !== nextSelected) {
        this.selectedContext = nextSelected;
        this.expandedRows.clear();
        this.persistState();
      }
    }
  }
  getEventsForSelectedContext() {
    if (this.selectedContext === "all-agents") {
      return this.flattenedEvents;
    }
    return this.agentEvents.get(this.selectedContext) ?? [];
  }
  filterEvents(events) {
    const query = this.eventFilterText.trim().toLowerCase();
    return events.filter((event) => {
      if (this.eventTypeFilter !== "all" && event.type !== this.eventTypeFilter) {
        return false;
      }
      if (!query) {
        return true;
      }
      const payloadText = this.stringifyPayload(event.payload, false).toLowerCase();
      return event.type.toLowerCase().includes(query) || event.agentId.toLowerCase().includes(query) || payloadText.includes(query);
    });
  }
  getLatestStateForAgent(agentId) {
    if (this.agentStates.has(agentId)) {
      const value = this.agentStates.get(agentId);
      return value === void 0 ? null : value;
    }
    const events = this.agentEvents.get(agentId) ?? [];
    const stateEvent = events.find((e) => e.type === "STATE_SNAPSHOT");
    if (!stateEvent) {
      return null;
    }
    return stateEvent.payload;
  }
  getLatestMessagesForAgent(agentId) {
    const messages = this.agentMessages.get(agentId);
    return messages ?? null;
  }
  getAgentStatus(agentId) {
    const events = this.agentEvents.get(agentId) ?? [];
    if (events.length === 0) {
      return "idle";
    }
    const runEvent = events.find((e) => e.type === "RUN_STARTED" || e.type === "RUN_FINISHED" || e.type === "RUN_ERROR");
    if (!runEvent) {
      return "idle";
    }
    if (runEvent.type === "RUN_ERROR") {
      return "error";
    }
    if (runEvent.type === "RUN_STARTED") {
      const finishedAfter = events.find(
        (e) => e.type === "RUN_FINISHED" && e.timestamp > runEvent.timestamp
      );
      return finishedAfter ? "idle" : "running";
    }
    return "idle";
  }
  getAgentStats(agentId) {
    const events = this.agentEvents.get(agentId) ?? [];
    const messages = this.agentMessages.get(agentId);
    const toolCallCount = messages ? messages.reduce((count, message) => count + (message.toolCalls?.length ?? 0), 0) : events.filter((e) => e.type === "TOOL_CALL_END").length;
    const messageCount = messages?.length ?? 0;
    return {
      totalEvents: events.length,
      lastActivity: events[0]?.timestamp ?? null,
      messages: messageCount,
      toolCalls: toolCallCount,
      errors: events.filter((e) => e.type === "RUN_ERROR").length
    };
  }
  renderToolCallDetails(toolCalls) {
    if (!Array.isArray(toolCalls) || toolCalls.length === 0) {
      return nothing;
    }
    return html`
      <div class="mt-2 space-y-2">
        ${toolCalls.map((call, index) => {
      const functionName = call.function?.name ?? call.toolName ?? "Unknown function";
      const callId = typeof call?.id === "string" ? call.id : `tool-call-${index + 1}`;
      const argsString = this.formatToolCallArguments(call.function?.arguments);
      return html`
            <div class="rounded-md border border-gray-200 bg-gray-50 p-3 text-xs text-gray-700">
              <div class="flex flex-wrap items-center justify-between gap-1 font-medium text-gray-900">
                <span>${functionName}</span>
                <span class="text-[10px] text-gray-500">ID: ${callId}</span>
              </div>
              ${argsString ? html`<pre class="mt-2 overflow-auto rounded bg-white p-2 text-[11px] leading-relaxed text-gray-800">${argsString}</pre>` : nothing}
            </div>
          `;
    })}
      </div>
    `;
  }
  formatToolCallArguments(args) {
    if (args === void 0 || args === null || args === "") {
      return null;
    }
    if (typeof args === "string") {
      try {
        const parsed = JSON.parse(args);
        return JSON.stringify(parsed, null, 2);
      } catch {
        return args;
      }
    }
    if (typeof args === "object") {
      try {
        return JSON.stringify(args, null, 2);
      } catch {
        return String(args);
      }
    }
    return String(args);
  }
  hasRenderableState(state) {
    if (state === null || state === void 0) {
      return false;
    }
    if (Array.isArray(state)) {
      return state.length > 0;
    }
    if (typeof state === "object") {
      return Object.keys(state).length > 0;
    }
    if (typeof state === "string") {
      const trimmed = state.trim();
      return trimmed.length > 0 && trimmed !== "{}";
    }
    return true;
  }
  formatStateForDisplay(state) {
    if (state === null || state === void 0) {
      return "";
    }
    if (typeof state === "string") {
      const trimmed = state.trim();
      if (trimmed.length === 0) {
        return "";
      }
      try {
        const parsed = JSON.parse(trimmed);
        return JSON.stringify(parsed, null, 2);
      } catch {
        return state;
      }
    }
    if (typeof state === "object") {
      try {
        return JSON.stringify(state, null, 2);
      } catch {
        return String(state);
      }
    }
    return String(state);
  }
  getEventBadgeClasses(type) {
    const base = "font-mono text-[10px] font-medium inline-flex items-center rounded-sm px-1.5 py-0.5 border";
    if (type.startsWith("RUN_")) {
      return `${base} bg-blue-50 text-blue-700 border-blue-200`;
    }
    if (type.startsWith("TEXT_MESSAGE")) {
      return `${base} bg-emerald-50 text-emerald-700 border-emerald-200`;
    }
    if (type.startsWith("TOOL_CALL")) {
      return `${base} bg-amber-50 text-amber-700 border-amber-200`;
    }
    if (type.startsWith("STATE")) {
      return `${base} bg-violet-50 text-violet-700 border-violet-200`;
    }
    if (type.startsWith("MESSAGES")) {
      return `${base} bg-sky-50 text-sky-700 border-sky-200`;
    }
    if (type === "RUN_ERROR") {
      return `${base} bg-rose-50 text-rose-700 border-rose-200`;
    }
    return `${base} bg-gray-100 text-gray-600 border-gray-200`;
  }
  stringifyPayload(payload, pretty) {
    try {
      if (payload === void 0) {
        return pretty ? "undefined" : "undefined";
      }
      if (typeof payload === "string") {
        return payload;
      }
      return JSON.stringify(payload, null, pretty ? 2 : 0) ?? "";
    } catch (error) {
      console.warn("Failed to stringify inspector payload", error);
      return String(payload);
    }
  }
  extractEventFromPayload(payload) {
    if (payload && typeof payload === "object" && "event" in payload) {
      return payload.event;
    }
    return payload;
  }
  async copyToClipboard(text, eventId) {
    try {
      await navigator.clipboard.writeText(text);
      this.copiedEvents.add(eventId);
      this.requestUpdate();
      setTimeout(() => {
        this.copiedEvents.delete(eventId);
        this.requestUpdate();
      }, 2e3);
    } catch (err) {
      console.error("Failed to copy to clipboard:", err);
    }
  }
  static {
    this.styles = [
      unsafeCSS(generated_default),
      css`
      :host {
        position: fixed;
        top: 0;
        left: 0;
        z-index: 2147483646;
        display: block;
        will-change: transform;
      }

      :host([data-transitioning="true"]) {
        transition: transform 300ms ease;
      }

      .console-button {
        transition:
          transform 300ms cubic-bezier(0.34, 1.56, 0.64, 1),
          opacity 160ms ease;
      }

      .console-button[data-dragging="true"] {
        transition: opacity 160ms ease;
      }

      .inspector-window[data-transitioning="true"] {
        transition: width 300ms ease, height 300ms ease;
      }

      .inspector-window[data-docked="true"] {
        border-radius: 0 !important;
        box-shadow: none !important;
      }

      .resize-handle {
        touch-action: none;
        user-select: none;
      }

      .dock-resize-handle {
        position: absolute;
        top: 0;
        right: 0;
        width: 10px;
        height: 100%;
        cursor: ew-resize;
        touch-action: none;
        z-index: 50;
        background: transparent;
      }

      .tooltip-target {
        position: relative;
      }

      .tooltip-target::after {
        content: attr(data-tooltip);
        position: absolute;
        top: calc(100% + 6px);
        left: 50%;
        transform: translateX(-50%) translateY(-4px);
        white-space: nowrap;
        background: rgba(17, 24, 39, 0.95);
        color: white;
        padding: 4px 8px;
        border-radius: 6px;
        font-size: 10px;
        line-height: 1.2;
        box-shadow: 0 4px 10px rgba(0, 0, 0, 0.15);
        opacity: 0;
        pointer-events: none;
        transition: opacity 120ms ease, transform 120ms ease;
        z-index: 4000;
      }

      .tooltip-target:hover::after {
        opacity: 1;
        transform: translateX(-50%) translateY(0);
      }

      .announcement-preview {
        position: absolute;
        top: 50%;
        transform: translateY(-50%);
        min-width: 300px;
        max-width: 300px;
        background: white;
        color: #111827;
        font-size: 13px;
        line-height: 1.4;
        border-radius: 12px;
        box-shadow: 0 12px 28px rgba(15, 23, 42, 0.22);
        padding: 10px 12px;
        display: inline-flex;
        align-items: flex-start;
        gap: 8px;
        z-index: 4500;
        animation: fade-slide-in 160ms ease;
        border: 1px solid rgba(148, 163, 184, 0.35);
        white-space: normal;
        word-break: break-word;
        text-align: left;
      }

      .announcement-preview[data-side="left"] {
        right: 100%;
        margin-right: 10px;
      }

      .announcement-preview[data-side="right"] {
        left: 100%;
        margin-left: 10px;
      }

      .announcement-preview__arrow {
        position: absolute;
        width: 10px;
        height: 10px;
        background: white;
        border: 1px solid rgba(148, 163, 184, 0.35);
        transform: rotate(45deg);
        top: 50%;
        margin-top: -5px;
        z-index: -1;
      }

      .announcement-preview[data-side="left"] .announcement-preview__arrow {
        right: -5px;
        box-shadow: 6px -6px 10px rgba(15, 23, 42, 0.12);
      }

      .announcement-preview[data-side="right"] .announcement-preview__arrow {
        left: -5px;
        box-shadow: -6px 6px 10px rgba(15, 23, 42, 0.12);
      }

      .announcement-dismiss {
        color: #6b7280;
        font-size: 12px;
        padding: 2px 8px;
        border-radius: 8px;
        border: 1px solid rgba(148, 163, 184, 0.5);
        background: rgba(248, 250, 252, 0.9);
        transition: background 120ms ease, color 120ms ease;
      }

      .announcement-dismiss:hover {
        background: rgba(241, 245, 249, 1);
        color: #111827;
      }

      .announcement-content {
        color: #111827;
        font-size: 14px;
        line-height: 1.6;
      }

      .announcement-content h1,
      .announcement-content h2,
      .announcement-content h3 {
        font-weight: 700;
        margin: 0.4rem 0 0.2rem;
      }

      .announcement-content h1 {
        font-size: 1.1rem;
      }

      .announcement-content h2 {
        font-size: 1rem;
      }

      .announcement-content h3 {
        font-size: 0.95rem;
      }

      .announcement-content p {
        margin: 0.25rem 0;
      }

      .announcement-content ul {
        list-style: disc;
        padding-left: 1.25rem;
        margin: 0.3rem 0;
      }

      .announcement-content ol {
        list-style: decimal;
        padding-left: 1.25rem;
        margin: 0.3rem 0;
      }

      .announcement-content a {
        color: #0f766e;
        text-decoration: underline;
      }
    `
    ];
  }
  connectedCallback() {
    super.connectedCallback();
    if (typeof window !== "undefined") {
      window.addEventListener("resize", this.handleResize);
      window.addEventListener("pointerdown", this.handleGlobalPointerDown);
      this.hydrateStateFromStorageEarly();
      this.tryAutoAttachCore();
      this.ensureAnnouncementLoading();
    }
  }
  disconnectedCallback() {
    super.disconnectedCallback();
    if (typeof window !== "undefined") {
      window.removeEventListener("resize", this.handleResize);
      window.removeEventListener("pointerdown", this.handleGlobalPointerDown);
    }
    this.removeDockStyles();
    this.detachFromCore();
  }
  firstUpdated() {
    if (typeof window === "undefined") {
      return;
    }
    if (!this._core) {
      this.tryAutoAttachCore();
    }
    this.measureContext("button");
    this.measureContext("window");
    this.contextState.button.anchor = { horizontal: "right", vertical: "top" };
    this.contextState.button.anchorOffset = { x: EDGE_MARGIN, y: EDGE_MARGIN };
    this.contextState.window.anchor = { horizontal: "right", vertical: "top" };
    this.contextState.window.anchorOffset = { x: EDGE_MARGIN, y: EDGE_MARGIN };
    this.hydrateStateFromStorage();
    if (this.isOpen && this.dockMode !== "floating") {
      this.applyDockStyles(true);
    }
    this.applyAnchorPosition("button");
    if (this.dockMode === "floating") {
      if (this.hasCustomPosition.window) {
        this.applyAnchorPosition("window");
      } else {
        this.centerContext("window");
      }
    }
    this.ensureAnnouncementLoading();
    this.updateHostTransform(this.isOpen ? "window" : "button");
  }
  render() {
    return this.isOpen ? this.renderWindow() : this.renderButton();
  }
  renderButton() {
    const buttonClasses = [
      "console-button",
      "group",
      "relative",
      "pointer-events-auto",
      "inline-flex",
      "h-12",
      "w-12",
      "items-center",
      "justify-center",
      "rounded-full",
      "border",
      "border-white/20",
      "bg-slate-950/95",
      "text-xs",
      "font-medium",
      "text-white",
      "ring-1",
      "ring-white/10",
      "backdrop-blur-md",
      "transition",
      "hover:border-white/30",
      "hover:bg-slate-900/95",
      "hover:scale-105",
      "focus-visible:outline",
      "focus-visible:outline-2",
      "focus-visible:outline-offset-2",
      "focus-visible:outline-rose-500",
      "touch-none",
      "select-none",
      this.isDragging ? "cursor-grabbing" : "cursor-grab"
    ].join(" ");
    return html`
      <button
        class=${buttonClasses}
        type="button"
        aria-label="Web Inspector"
        data-drag-context="button"
        data-dragging=${this.isDragging && this.pointerContext === "button" ? "true" : "false"}
        @pointerdown=${this.handlePointerDown}
        @pointermove=${this.handlePointerMove}
        @pointerup=${this.handlePointerUp}
        @pointercancel=${this.handlePointerCancel}
        @click=${this.handleButtonClick}
      >
        ${this.renderAnnouncementPreview()}
        <img src=${inspector_logo_icon_default} alt="Inspector logo" class="h-5 w-auto" loading="lazy" />
      </button>
    `;
  }
  renderWindow() {
    const windowState = this.contextState.window;
    const isDocked = this.dockMode !== "floating";
    const isTransitioning = this.hasAttribute("data-transitioning");
    const windowStyles = isDocked ? this.getDockedWindowStyles() : {
      width: `${Math.round(windowState.size.width)}px`,
      height: `${Math.round(windowState.size.height)}px`,
      minWidth: `${MIN_WINDOW_WIDTH}px`,
      minHeight: `${MIN_WINDOW_HEIGHT}px`
    };
    const hasContextDropdown = this.contextOptions.length > 0;
    const contextDropdown = hasContextDropdown ? this.renderContextDropdown() : nothing;
    const coreStatus = this.getCoreStatusSummary();
    const agentSelector = hasContextDropdown ? contextDropdown : html`
          <div class="flex items-center gap-2 rounded-md border border-dashed border-gray-200 px-2 py-1 text-xs text-gray-400">
            <span>${this.renderIcon("Bot")}</span>
            <span class="truncate">No agents available</span>
          </div>
        `;
    return html`
      <section
        class="inspector-window pointer-events-auto relative flex flex-col overflow-hidden rounded-xl border border-gray-200 bg-white text-gray-900 shadow-lg"
        style=${styleMap(windowStyles)}
        data-docked=${isDocked}
        data-transitioning=${isTransitioning}
      >
        ${isDocked ? html`
              <div
                class="dock-resize-handle pointer-events-auto"
                role="presentation"
                aria-hidden="true"
                @pointerdown=${this.handleResizePointerDown}
                @pointermove=${this.handleResizePointerMove}
                @pointerup=${this.handleResizePointerUp}
                @pointercancel=${this.handleResizePointerCancel}
              ></div>
            ` : nothing}
        <div class="flex flex-1 flex-col overflow-hidden bg-white text-gray-800">
          <div
            class="drag-handle relative z-30 flex flex-col border-b border-gray-200 bg-white/95 backdrop-blur-sm ${isDocked ? "" : this.isDragging && this.pointerContext === "window" ? "cursor-grabbing" : "cursor-grab"}"
            data-drag-context="window"
            @pointerdown=${isDocked ? void 0 : this.handlePointerDown}
            @pointermove=${isDocked ? void 0 : this.handlePointerMove}
            @pointerup=${isDocked ? void 0 : this.handlePointerUp}
            @pointercancel=${isDocked ? void 0 : this.handlePointerCancel}
          >
            <div class="flex flex-wrap items-center gap-3 px-4 py-3">
              <div class="flex items-center min-w-0">
                <img src=${inspector_logo_default} alt="Inspector logo" class="h-6 w-auto" loading="lazy" />
              </div>
              <div class="ml-auto flex min-w-0 items-center gap-2">
                <div class="min-w-[160px] max-w-xs">
                  ${agentSelector}
                </div>
                <div class="flex items-center gap-1">
                  ${this.renderDockControls()}
                  <button
                    class="flex h-8 w-8 items-center justify-center rounded-md text-gray-400 transition hover:bg-gray-100 hover:text-gray-600 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-gray-400"
                    type="button"
                    aria-label="Close Web Inspector"
                    @pointerdown=${this.handleClosePointerDown}
                    @click=${this.handleCloseClick}
                  >
                    ${this.renderIcon("X")}
                  </button>
                </div>
              </div>
            </div>
            <div class="flex flex-wrap items-center gap-2 border-t border-gray-100 px-3 py-2 text-xs">
              ${this.menuItems.map(({ key, label, icon }) => {
      const isSelected = this.selectedMenu === key;
      const tabClasses = [
        "inline-flex items-center gap-2 rounded-md px-3 py-2 transition focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-gray-300",
        isSelected ? "bg-gray-900 text-white shadow-sm" : "text-gray-600 hover:bg-gray-100 hover:text-gray-900"
      ].join(" ");
      return html`
                  <button
                    type="button"
                    class=${tabClasses}
                    aria-pressed=${isSelected}
                    @click=${() => this.handleMenuSelect(key)}
                  >
                    <span class="text-gray-400 ${isSelected ? "text-white" : ""}">
                      ${this.renderIcon(icon)}
                    </span>
                    <span>${label}</span>
                  </button>
                `;
    })}
            </div>
          </div>
            <div class="flex flex-1 flex-col overflow-hidden">
              <div class="flex-1 overflow-auto">
                ${this.renderAnnouncementPanel()}
                ${this.renderCoreWarningBanner()}
                ${this.renderMainContent()}
                <slot></slot>
              </div>
              <div class="border-t border-gray-200 bg-gray-50 px-4 py-2">
                <div
                  class="flex items-center gap-2 rounded-md px-3 py-2 text-xs ${coreStatus.tone} w-full overflow-hidden my-1"
                  title=${coreStatus.description}
                >
                  <span class="flex h-6 w-6 items-center justify-center rounded bg-white/60">
                    ${this.renderIcon("Activity")}
                  </span>
                  <span class="font-medium">${coreStatus.label}</span>
                  <span class="truncate text-[11px] opacity-80">${coreStatus.description}</span>
                </div>
              </div>
            </div>
          </div>
        <div
          class="resize-handle pointer-events-auto absolute bottom-1 right-1 flex h-5 w-5 cursor-nwse-resize items-center justify-center text-gray-400 transition hover:text-gray-600"
          role="presentation"
          aria-hidden="true"
          @pointerdown=${this.handleResizePointerDown}
          @pointermove=${this.handleResizePointerMove}
          @pointerup=${this.handleResizePointerUp}
          @pointercancel=${this.handleResizePointerCancel}
        >
          <svg
            class="h-3 w-3"
            viewBox="0 0 16 16"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-width="1.5"
          >
            <path d="M5 15L15 5" />
            <path d="M9 15L15 9" />
          </svg>
        </div>
      </section>
    `;
  }
  hydrateStateFromStorageEarly() {
    if (typeof document === "undefined" || typeof window === "undefined") {
      return;
    }
    const persisted = loadInspectorState(INSPECTOR_STORAGE_KEY);
    if (!persisted) {
      return;
    }
    if (typeof persisted.isOpen === "boolean") {
      this.isOpen = persisted.isOpen;
    }
    if (isValidDockMode(persisted.dockMode)) {
      this.dockMode = persisted.dockMode;
    }
    if (typeof persisted.selectedMenu === "string") {
      const validMenu = this.menuItems.find((item) => item.key === persisted.selectedMenu);
      if (validMenu) {
        this.selectedMenu = validMenu.key;
      }
    }
    if (typeof persisted.selectedContext === "string") {
      this.selectedContext = persisted.selectedContext;
      this.pendingSelectedContext = persisted.selectedContext;
    }
  }
  hydrateStateFromStorage() {
    if (typeof document === "undefined" || typeof window === "undefined") {
      return;
    }
    const persisted = loadInspectorState(INSPECTOR_STORAGE_KEY);
    if (!persisted) {
      return;
    }
    const persistedButton = persisted.button;
    if (persistedButton) {
      if (isValidAnchor(persistedButton.anchor)) {
        this.contextState.button.anchor = persistedButton.anchor;
      }
      if (isValidPosition(persistedButton.anchorOffset)) {
        this.contextState.button.anchorOffset = persistedButton.anchorOffset;
      }
      if (typeof persistedButton.hasCustomPosition === "boolean") {
        this.hasCustomPosition.button = persistedButton.hasCustomPosition;
      }
    }
    const persistedWindow = persisted.window;
    if (persistedWindow) {
      if (isValidAnchor(persistedWindow.anchor)) {
        this.contextState.window.anchor = persistedWindow.anchor;
      }
      if (isValidPosition(persistedWindow.anchorOffset)) {
        this.contextState.window.anchorOffset = persistedWindow.anchorOffset;
      }
      if (isValidSize(persistedWindow.size)) {
        this.contextState.window.size = this.clampWindowSize(persistedWindow.size);
      }
      if (typeof persistedWindow.hasCustomPosition === "boolean") {
        this.hasCustomPosition.window = persistedWindow.hasCustomPosition;
      }
    }
    if (typeof persisted.selectedContext === "string") {
      this.selectedContext = persisted.selectedContext;
      this.pendingSelectedContext = persisted.selectedContext;
    }
  }
  get activeContext() {
    return this.isOpen ? "window" : "button";
  }
  measureContext(context) {
    const selector = context === "window" ? ".inspector-window" : ".console-button";
    const element = this.renderRoot?.querySelector(selector);
    if (!element) {
      return;
    }
    const fallback = context === "window" ? DEFAULT_WINDOW_SIZE : DEFAULT_BUTTON_SIZE;
    updateSizeFromElement(this.contextState[context], element, fallback);
  }
  centerContext(context) {
    if (typeof window === "undefined") {
      return;
    }
    const viewport = this.getViewportSize();
    centerContext(this.contextState[context], viewport, EDGE_MARGIN);
    if (context === this.activeContext) {
      this.updateHostTransform(context);
    }
    this.hasCustomPosition[context] = false;
    this.persistState();
  }
  ensureWindowPlacement() {
    if (typeof window === "undefined") {
      return;
    }
    if (!this.hasCustomPosition.window) {
      this.centerContext("window");
      return;
    }
    const viewport = this.getViewportSize();
    keepPositionWithinViewport(this.contextState.window, viewport, EDGE_MARGIN);
    updateAnchorFromPosition(this.contextState.window, viewport, EDGE_MARGIN);
    this.updateHostTransform("window");
    this.persistState();
  }
  constrainToViewport(position, context) {
    if (typeof window === "undefined") {
      return position;
    }
    const viewport = this.getViewportSize();
    return constrainToViewport(this.contextState[context], position, viewport, EDGE_MARGIN);
  }
  keepPositionWithinViewport(context) {
    if (typeof window === "undefined") {
      return;
    }
    const viewport = this.getViewportSize();
    keepPositionWithinViewport(this.contextState[context], viewport, EDGE_MARGIN);
  }
  getViewportSize() {
    if (typeof window === "undefined") {
      return { ...DEFAULT_WINDOW_SIZE };
    }
    return { width: window.innerWidth, height: window.innerHeight };
  }
  persistState() {
    const state = {
      button: {
        anchor: this.contextState.button.anchor,
        anchorOffset: this.contextState.button.anchorOffset,
        hasCustomPosition: this.hasCustomPosition.button
      },
      window: {
        anchor: this.contextState.window.anchor,
        anchorOffset: this.contextState.window.anchorOffset,
        size: {
          width: Math.round(this.contextState.window.size.width),
          height: Math.round(this.contextState.window.size.height)
        },
        hasCustomPosition: this.hasCustomPosition.window
      },
      isOpen: this.isOpen,
      dockMode: this.dockMode,
      selectedMenu: this.selectedMenu,
      selectedContext: this.selectedContext
    };
    saveInspectorState(INSPECTOR_STORAGE_KEY, state);
    this.pendingSelectedContext = state.selectedContext ?? null;
  }
  clampWindowSize(size) {
    const minWidth = this.dockMode === "docked-left" ? MIN_WINDOW_WIDTH_DOCKED_LEFT : MIN_WINDOW_WIDTH;
    if (typeof window === "undefined") {
      return {
        width: Math.max(minWidth, size.width),
        height: Math.max(MIN_WINDOW_HEIGHT, size.height)
      };
    }
    const viewport = this.getViewportSize();
    return clampSize(size, viewport, EDGE_MARGIN, minWidth, MIN_WINDOW_HEIGHT);
  }
  setDockMode(mode) {
    if (this.dockMode === mode) {
      return;
    }
    this.startHostTransition();
    this.removeDockStyles();
    this.dockMode = mode;
    if (mode !== "floating") {
      if (mode === "docked-left") {
        this.contextState.window.size.width = DOCKED_LEFT_WIDTH;
      }
      this.applyDockStyles();
    } else {
      this.contextState.window.size = { ...DEFAULT_WINDOW_SIZE };
      this.centerContext("window");
    }
    this.persistState();
    this.requestUpdate();
    this.updateHostTransform("window");
  }
  startHostTransition(duration = 300) {
    this.setAttribute("data-transitioning", "true");
    if (this.transitionTimeoutId !== null) {
      clearTimeout(this.transitionTimeoutId);
    }
    this.transitionTimeoutId = setTimeout(() => {
      this.removeAttribute("data-transitioning");
      this.transitionTimeoutId = null;
    }, duration);
  }
  applyDockStyles(skipTransition = false) {
    if (typeof document === "undefined" || !document.body) {
      return;
    }
    const computedStyle = window.getComputedStyle(document.body);
    this.previousBodyMargins = {
      left: computedStyle.marginLeft,
      bottom: computedStyle.marginBottom
    };
    if (!this.isResizing && !skipTransition) {
      document.body.style.transition = "margin 300ms ease";
    }
    if (this.dockMode === "docked-left") {
      document.body.style.marginLeft = `${this.contextState.window.size.width}px`;
    }
    if (!this.isResizing && !skipTransition) {
      setTimeout(() => {
        if (document.body) {
          document.body.style.transition = "";
        }
      }, 300);
    }
  }
  removeDockStyles() {
    if (typeof document === "undefined" || !document.body) {
      return;
    }
    if (!this.isResizing) {
      document.body.style.transition = "margin 300ms ease";
    }
    if (this.previousBodyMargins) {
      document.body.style.marginLeft = this.previousBodyMargins.left;
      document.body.style.marginBottom = this.previousBodyMargins.bottom;
      this.previousBodyMargins = null;
    } else {
      document.body.style.marginLeft = "";
      document.body.style.marginBottom = "";
    }
    setTimeout(() => {
      if (document.body) {
        document.body.style.transition = "";
      }
    }, 300);
  }
  updateHostTransform(context = this.activeContext) {
    if (context !== this.activeContext) {
      return;
    }
    if (this.isOpen && this.dockMode === "docked-left") {
      this.style.transform = `translate3d(0, 0, 0)`;
    } else {
      const { position } = this.contextState[context];
      this.style.transform = `translate3d(${position.x}px, ${position.y}px, 0)`;
    }
  }
  setDragging(value) {
    if (this.isDragging !== value) {
      this.isDragging = value;
      this.requestUpdate();
    }
  }
  updateAnchorFromPosition(context) {
    if (typeof window === "undefined") {
      return;
    }
    const viewport = this.getViewportSize();
    updateAnchorFromPosition(this.contextState[context], viewport, EDGE_MARGIN);
  }
  snapButtonToCorner() {
    if (typeof window === "undefined") {
      return;
    }
    const viewport = this.getViewportSize();
    const state = this.contextState.button;
    const centerX = state.position.x + state.size.width / 2;
    const centerY = state.position.y + state.size.height / 2;
    const horizontal = centerX < viewport.width / 2 ? "left" : "right";
    const vertical = centerY < viewport.height / 2 ? "top" : "bottom";
    state.anchor = { horizontal, vertical };
    state.anchorOffset = { x: EDGE_MARGIN, y: EDGE_MARGIN };
    this.startHostTransition();
    this.applyAnchorPosition("button");
  }
  applyAnchorPosition(context) {
    if (typeof window === "undefined") {
      return;
    }
    const viewport = this.getViewportSize();
    applyAnchorPosition(this.contextState[context], viewport, EDGE_MARGIN);
    this.updateHostTransform(context);
    this.persistState();
  }
  resetResizeTracking() {
    this.resizePointerId = null;
    this.resizeStart = null;
    this.resizeInitialSize = null;
    this.isResizing = false;
  }
  resetPointerTracking() {
    this.pointerId = null;
    this.dragStart = null;
    this.pointerContext = null;
    this.setDragging(false);
    this.draggedDuringInteraction = false;
  }
  openInspector() {
    if (this.isOpen) {
      return;
    }
    this.showAnnouncementPreview = false;
    this.ensureAnnouncementLoading();
    this.isOpen = true;
    this.persistState();
    if (this.dockMode !== "floating") {
      this.applyDockStyles();
    }
    this.ensureWindowPlacement();
    this.requestUpdate();
    void this.updateComplete.then(() => {
      this.measureContext("window");
      if (this.dockMode === "floating") {
        if (this.hasCustomPosition.window) {
          this.applyAnchorPosition("window");
        } else {
          this.centerContext("window");
        }
      } else {
        this.updateHostTransform("window");
      }
    });
  }
  closeInspector() {
    if (!this.isOpen) {
      return;
    }
    this.isOpen = false;
    if (this.dockMode !== "floating") {
      this.removeDockStyles();
    }
    this.persistState();
    this.updateHostTransform("button");
    this.requestUpdate();
    void this.updateComplete.then(() => {
      this.measureContext("button");
      this.applyAnchorPosition("button");
    });
  }
  renderIcon(name) {
    const iconNode = icons[name];
    if (!iconNode) {
      return nothing;
    }
    const svgAttrs = {
      xmlns: "http://www.w3.org/2000/svg",
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      "stroke-width": "1.5",
      "stroke-linecap": "round",
      "stroke-linejoin": "round",
      class: "h-3.5 w-3.5"
    };
    const svgMarkup = `<svg ${this.serializeAttributes(svgAttrs)}>${iconNode.map(([tag, attrs]) => `<${tag} ${this.serializeAttributes(attrs)} />`).join("")}</svg>`;
    return unsafeHTML(svgMarkup);
  }
  renderDockControls() {
    if (this.dockMode === "floating") {
      return html`
        <button
          class="flex h-8 w-8 items-center justify-center rounded-md text-gray-400 transition hover:bg-gray-100 hover:text-gray-600 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-gray-400"
          type="button"
          aria-label="Dock to left"
          title="Dock Left"
          @click=${() => this.handleDockClick("docked-left")}
        >
          ${this.renderIcon("PanelLeft")}
        </button>
      `;
    } else {
      return html`
        <button
          class="flex h-8 w-8 items-center justify-center rounded-md text-gray-400 transition hover:bg-gray-100 hover:text-gray-600 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-gray-400"
          type="button"
          aria-label="Float window"
          title="Float"
          @click=${() => this.handleDockClick("floating")}
        >
          ${this.renderIcon("Maximize2")}
        </button>
      `;
    }
  }
  getDockedWindowStyles() {
    if (this.dockMode === "docked-left") {
      return {
        position: "fixed",
        top: "0",
        left: "0",
        bottom: "0",
        width: `${Math.round(this.contextState.window.size.width)}px`,
        height: "100vh",
        minWidth: `${MIN_WINDOW_WIDTH_DOCKED_LEFT}px`,
        borderRadius: "0"
      };
    }
    return {
      width: `${Math.round(this.contextState.window.size.width)}px`,
      height: `${Math.round(this.contextState.window.size.height)}px`,
      minWidth: `${MIN_WINDOW_WIDTH}px`,
      minHeight: `${MIN_WINDOW_HEIGHT}px`
    };
  }
  handleDockClick(mode) {
    this.setDockMode(mode);
  }
  serializeAttributes(attributes) {
    return Object.entries(attributes).filter(([key, value]) => key !== "key" && value !== void 0 && value !== null && value !== "").map(([key, value]) => `${key}="${String(value).replace(/"/g, "&quot;")}"`).join(" ");
  }
  sanitizeForLogging(value, depth = 0, seen = /* @__PURE__ */ new WeakSet()) {
    if (value === void 0) {
      return "[undefined]";
    }
    if (value === null || typeof value === "number" || typeof value === "boolean") {
      return value;
    }
    if (typeof value === "string") {
      return value;
    }
    if (typeof value === "bigint" || typeof value === "symbol" || typeof value === "function") {
      return String(value);
    }
    if (value instanceof Date) {
      return value.toISOString();
    }
    if (Array.isArray(value)) {
      if (depth >= 4) {
        return "[Truncated depth]";
      }
      return value.map((item) => this.sanitizeForLogging(item, depth + 1, seen));
    }
    if (typeof value === "object") {
      if (seen.has(value)) {
        return "[Circular]";
      }
      seen.add(value);
      if (depth >= 4) {
        return "[Truncated depth]";
      }
      const result = {};
      for (const [key, entry] of Object.entries(value)) {
        result[key] = this.sanitizeForLogging(entry, depth + 1, seen);
      }
      return result;
    }
    return String(value);
  }
  normalizeEventPayload(_type, payload) {
    if (payload && typeof payload === "object" && "event" in payload) {
      const { event, ...rest } = payload;
      const cleaned = Object.keys(rest).length === 0 ? event : { event, ...rest };
      return this.sanitizeForLogging(cleaned);
    }
    return this.sanitizeForLogging(payload);
  }
  normalizeMessageContent(content) {
    if (typeof content === "string") {
      return content;
    }
    if (content && typeof content === "object" && "text" in content) {
      const maybeText = content.text;
      if (typeof maybeText === "string") {
        return maybeText;
      }
    }
    if (content === null || content === void 0) {
      return "";
    }
    if (typeof content === "object") {
      try {
        return JSON.stringify(this.sanitizeForLogging(content));
      } catch {
        return "";
      }
    }
    return String(content);
  }
  normalizeToolCalls(raw) {
    if (!Array.isArray(raw)) {
      return [];
    }
    return raw.map((entry) => {
      if (!entry || typeof entry !== "object") {
        return null;
      }
      const call = entry;
      const fn = call.function;
      const functionName = typeof fn?.name === "string" ? fn.name : typeof call.toolName === "string" ? call.toolName : void 0;
      const args = fn && "arguments" in fn ? fn.arguments : call.arguments;
      const normalized = {
        id: typeof call.id === "string" ? call.id : void 0,
        toolName: typeof call.toolName === "string" ? call.toolName : functionName,
        status: typeof call.status === "string" ? call.status : void 0
      };
      if (functionName) {
        normalized.function = {
          name: functionName,
          arguments: this.sanitizeForLogging(args)
        };
      }
      return normalized;
    }).filter((call) => Boolean(call));
  }
  normalizeAgentMessage(message) {
    if (!message || typeof message !== "object") {
      return null;
    }
    const raw = message;
    const role = typeof raw.role === "string" ? raw.role : "unknown";
    const contentText = this.normalizeMessageContent(raw.content);
    const toolCalls = this.normalizeToolCalls(raw.toolCalls);
    return {
      id: typeof raw.id === "string" ? raw.id : void 0,
      role,
      contentText,
      contentRaw: raw.content !== void 0 ? this.sanitizeForLogging(raw.content) : void 0,
      toolCalls
    };
  }
  normalizeAgentMessages(messages) {
    if (!Array.isArray(messages)) {
      return null;
    }
    const normalized = messages.map((message) => this.normalizeAgentMessage(message)).filter((msg) => msg !== null);
    return normalized;
  }
  normalizeContextStore(context) {
    if (!context || typeof context !== "object") {
      return {};
    }
    const normalized = {};
    for (const [key, entry] of Object.entries(context)) {
      if (entry && typeof entry === "object" && "value" in entry) {
        const candidate = entry;
        const description = typeof candidate.description === "string" && candidate.description.trim().length > 0 ? candidate.description : void 0;
        normalized[key] = { description, value: candidate.value };
      } else {
        normalized[key] = { value: entry };
      }
    }
    return normalized;
  }
  getSelectedMenu() {
    const found = this.menuItems.find((item) => item.key === this.selectedMenu);
    return found ?? this.menuItems[0];
  }
  renderCoreWarningBanner() {
    if (this._core) {
      return nothing;
    }
    return html`
      <div class="mx-4 my-3 flex items-start gap-2 rounded-md border border-amber-200 bg-amber-50 px-3 py-2 text-xs text-amber-800">
        <span class="mt-0.5 shrink-0 text-amber-600">${this.renderIcon("AlertTriangle")}</span>
        <div class="space-y-1">
          <div class="font-semibold text-amber-900">CopilotKit core not attached</div>
          <p class="text-[11px] leading-snug text-amber-800">
            Pass a live <code>CopilotKitCore</code> instance to <code>&lt;cpk-web-inspector&gt;</code> or expose it on
            <code>window.__COPILOTKIT_CORE__</code> for auto-attach.
          </p>
        </div>
      </div>
    `;
  }
  getCoreStatusSummary() {
    if (!this._core) {
      return {
        label: "Core not attached",
        tone: "border border-amber-200 bg-amber-50 text-amber-800",
        description: "Pass a CopilotKitCore instance to <cpk-web-inspector> or enable auto-attach."
      };
    }
    const status = this.runtimeStatus ?? CopilotKitCoreRuntimeConnectionStatus.Disconnected;
    const lastErrorMessage = this.lastCoreError?.message;
    if (status === CopilotKitCoreRuntimeConnectionStatus.Error) {
      return {
        label: "Runtime error",
        tone: "border border-rose-200 bg-rose-50 text-rose-700",
        description: lastErrorMessage ?? "CopilotKit runtime reported an error."
      };
    }
    if (status === CopilotKitCoreRuntimeConnectionStatus.Connecting) {
      return {
        label: "Connecting",
        tone: "border border-amber-200 bg-amber-50 text-amber-800",
        description: "Waiting for CopilotKit runtime to finish connecting."
      };
    }
    if (status === CopilotKitCoreRuntimeConnectionStatus.Connected) {
      return {
        label: "Connected",
        tone: "border border-emerald-200 bg-emerald-50 text-emerald-700",
        description: "Live runtime connection established."
      };
    }
    return {
      label: "Disconnected",
      tone: "border border-gray-200 bg-gray-50 text-gray-700",
      description: lastErrorMessage ?? "Waiting for CopilotKit runtime to connect."
    };
  }
  renderMainContent() {
    if (this.selectedMenu === "ag-ui-events") {
      return this.renderEventsTable();
    }
    if (this.selectedMenu === "agents") {
      return this.renderAgentsView();
    }
    if (this.selectedMenu === "frontend-tools") {
      return this.renderToolsView();
    }
    if (this.selectedMenu === "agent-context") {
      return this.renderContextView();
    }
    return nothing;
  }
  renderEventsTable() {
    const events = this.getEventsForSelectedContext();
    const filteredEvents = this.filterEvents(events);
    const selectedLabel = this.selectedContext === "all-agents" ? "all agents" : `agent ${this.selectedContext}`;
    if (events.length === 0) {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-center">
          <div class="max-w-md">
            <div class="mb-3 flex justify-center text-gray-300 [&>svg]:!h-8 [&>svg]:!w-8">
              ${this.renderIcon("Zap")}
            </div>
            <p class="text-sm text-gray-600">No events yet</p>
            <p class="mt-2 text-xs text-gray-500">Trigger an agent run to see live activity.</p>
          </div>
        </div>
      `;
    }
    if (filteredEvents.length === 0) {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-center">
          <div class="max-w-md space-y-3">
            <div class="flex justify-center text-gray-300 [&>svg]:!h-8 [&>svg]:!w-8">
              ${this.renderIcon("Filter")}
            </div>
            <p class="text-sm text-gray-600">No events match the current filters.</p>
            <div>
              <button
                type="button"
                class="inline-flex items-center gap-1 rounded-md bg-gray-900 px-3 py-1.5 text-[11px] font-medium text-white transition hover:bg-gray-800"
                @click=${this.resetEventFilters}
              >
                ${this.renderIcon("RefreshCw")}
                <span>Reset filters</span>
              </button>
            </div>
          </div>
        </div>
      `;
    }
    return html`
      <div class="flex h-full flex-col">
        <div class="flex flex-col gap-1.5 border-b border-gray-200 bg-white px-4 py-2.5">
          <div class="flex flex-wrap items-center gap-2">
            <div class="relative min-w-[200px] flex-1">
              <input
                type="search"
                class="w-full rounded-md border border-gray-200 px-3 py-1.5 text-[11px] text-gray-700 shadow-sm outline-none ring-1 ring-transparent transition focus:border-gray-300 focus:ring-gray-200"
                placeholder="Search agent, type, payload"
                .value=${this.eventFilterText}
                @input=${this.handleEventFilterInput}
              />
            </div>
            <select
              class="w-40 rounded-md border border-gray-200 bg-white px-2 py-1.5 text-[11px] text-gray-700 shadow-sm outline-none transition focus:border-gray-300 focus:ring-2 focus:ring-gray-200"
              .value=${this.eventTypeFilter}
              @change=${this.handleEventTypeChange}
            >
              <option value="all">All event types</option>
              ${AGENT_EVENT_TYPES.map(
      (type) => html`<option value=${type}>${type.toLowerCase().replace(/_/g, " ")}</option>`
    )}
            </select>
            <div class="flex items-center gap-1 text-[11px]">
              <button
                type="button"
                class="tooltip-target flex h-8 w-8 items-center justify-center rounded-md border border-gray-200 bg-white text-gray-600 transition hover:bg-gray-50 disabled:cursor-not-allowed disabled:opacity-50"
                title="Reset filters"
                data-tooltip="Reset filters"
                aria-label="Reset filters"
                @click=${this.resetEventFilters}
                ?disabled=${!this.eventFilterText && this.eventTypeFilter === "all"}
              >
                ${this.renderIcon("RotateCw")}
              </button>
              <button
                type="button"
                class="tooltip-target flex h-8 w-8 items-center justify-center rounded-md border border-gray-200 bg-white text-gray-600 transition hover:bg-gray-50 disabled:cursor-not-allowed disabled:opacity-50"
                title="Export JSON"
                data-tooltip="Export JSON"
                aria-label="Export JSON"
                @click=${() => this.exportEvents(filteredEvents)}
                ?disabled=${filteredEvents.length === 0}
              >
                ${this.renderIcon("Download")}
              </button>
              <button
                type="button"
                class="tooltip-target flex h-8 w-8 items-center justify-center rounded-md border border-gray-200 bg-white text-gray-600 transition hover:bg-gray-50 disabled:cursor-not-allowed disabled:opacity-50"
                title="Clear events"
                data-tooltip="Clear events"
                aria-label="Clear events"
                @click=${this.handleClearEvents}
                ?disabled=${events.length === 0}
              >
                ${this.renderIcon("Trash2")}
              </button>
            </div>
          </div>
          <div class="text-[11px] text-gray-500">
            Showing ${filteredEvents.length} of ${events.length}${this.selectedContext === "all-agents" ? "" : ` for ${selectedLabel}`}
          </div>
        </div>
        <div class="relative h-full w-full overflow-y-auto overflow-x-hidden">
          <table class="w-full table-fixed border-collapse text-xs box-border">
            <thead class="sticky top-0 z-10">
              <tr class="bg-white">
                <th class="border-b border-gray-200 bg-white px-3 py-2 text-left font-medium text-gray-900">
                  Agent
                </th>
                <th class="border-b border-gray-200 bg-white px-3 py-2 text-left font-medium text-gray-900">
                  Time
                </th>
                <th class="border-b border-gray-200 bg-white px-3 py-2 text-left font-medium text-gray-900">
                  Event Type
                </th>
                <th class="border-b border-gray-200 bg-white px-3 py-2 text-left font-medium text-gray-900">
                  AG-UI Event
                </th>
              </tr>
            </thead>
            <tbody>
              ${filteredEvents.map((event, index) => {
      const rowBg = index % 2 === 0 ? "bg-white" : "bg-gray-50/50";
      const badgeClasses = this.getEventBadgeClasses(event.type);
      const extractedEvent = this.extractEventFromPayload(event.payload);
      const inlineEvent = this.stringifyPayload(extractedEvent, false) || "\u2014";
      const prettyEvent = this.stringifyPayload(extractedEvent, true) || inlineEvent;
      const isExpanded = this.expandedRows.has(event.id);
      return html`
                  <tr
                    class="${rowBg} cursor-pointer transition hover:bg-blue-50/50"
                    @click=${() => this.toggleRowExpansion(event.id)}
                  >
                    <td class="border-l border-r border-b border-gray-200 px-3 py-2">
                      <span class="font-mono text-[11px] text-gray-600">${event.agentId}</span>
                    </td>
                    <td class="border-r border-b border-gray-200 px-3 py-2 font-mono text-[11px] text-gray-600">
                      <span title=${new Date(event.timestamp).toLocaleString()}>
                        ${new Date(event.timestamp).toLocaleTimeString()}
                      </span>
                    </td>
                    <td class="border-r border-b border-gray-200 px-3 py-2">
                      <span class=${badgeClasses}>${event.type}</span>
                    </td>
                    <td class="border-r border-b border-gray-200 px-3 py-2 font-mono text-[10px] text-gray-600 ${isExpanded ? "" : "truncate max-w-xs"}">
                      ${isExpanded ? html`
                            <div class="group relative">
                              <pre class="m-0 whitespace-pre-wrap break-words text-[10px] font-mono text-gray-600">${prettyEvent}</pre>
                              <button
                                class="absolute right-0 top-0 cursor-pointer rounded px-2 py-1 text-[10px] opacity-0 transition group-hover:opacity-100 ${this.copiedEvents.has(event.id) ? "bg-green-100 text-green-700" : "bg-gray-100 text-gray-600 hover:bg-gray-200 hover:text-gray-900"}"
                                @click=${(e) => {
        e.stopPropagation();
        this.copyToClipboard(prettyEvent, event.id);
      }}
                              >
                                ${this.copiedEvents.has(event.id) ? html`<span> Copied</span>` : html`<span>Copy</span>`}
                              </button>
                            </div>
                          ` : inlineEvent}
                    </td>
                  </tr>
                `;
    })}
            </tbody>
          </table>
        </div>
      </div>
    `;
  }
  handleEventFilterInput(event) {
    const target = event.target;
    this.eventFilterText = target?.value ?? "";
    this.requestUpdate();
  }
  handleEventTypeChange(event) {
    const target = event.target;
    const value = target?.value;
    if (!value) {
      return;
    }
    this.eventTypeFilter = value;
    this.requestUpdate();
  }
  resetEventFilters() {
    this.eventFilterText = "";
    this.eventTypeFilter = "all";
    this.requestUpdate();
  }
  exportEvents(events) {
    try {
      const payload = JSON.stringify(events, null, 2);
      const blob = new Blob([payload], { type: "application/json" });
      const url = URL.createObjectURL(blob);
      const anchor = document.createElement("a");
      anchor.href = url;
      anchor.download = `copilotkit-events-${Date.now()}.json`;
      anchor.click();
      URL.revokeObjectURL(url);
    } catch (error) {
      console.error("Failed to export events", error);
    }
  }
  renderAgentsView() {
    if (this.selectedContext === "all-agents") {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-center">
          <div class="max-w-md">
            <div class="mb-3 flex justify-center text-gray-300 [&>svg]:!h-8 [&>svg]:!w-8">
              ${this.renderIcon("Bot")}
            </div>
            <p class="text-sm text-gray-600">No agent selected</p>
            <p class="mt-2 text-xs text-gray-500">Select an agent from the dropdown above to view details.</p>
          </div>
        </div>
      `;
    }
    const agentId = this.selectedContext;
    const status = this.getAgentStatus(agentId);
    const stats = this.getAgentStats(agentId);
    const state = this.getLatestStateForAgent(agentId);
    const messages = this.getLatestMessagesForAgent(agentId);
    const statusColors = {
      running: "bg-emerald-50 text-emerald-700",
      idle: "bg-gray-100 text-gray-600",
      error: "bg-rose-50 text-rose-700"
    };
    return html`
      <div class="flex flex-col gap-4 p-4 overflow-auto">
        <!-- Agent Overview Card -->
        <div class="rounded-lg border border-gray-200 bg-white p-4">
          <div class="flex items-start justify-between mb-4">
            <div class="flex items-center gap-3">
              <div class="flex h-10 w-10 items-center justify-center rounded-lg bg-blue-100 text-blue-600">
                ${this.renderIcon("Bot")}
              </div>
              <div>
                <h3 class="font-semibold text-sm text-gray-900">${agentId}</h3>
                <span class="inline-flex items-center gap-1.5 rounded-full px-2 py-0.5 text-xs font-medium ${statusColors[status]} relative -translate-y-[2px]">
                  <span class="h-1.5 w-1.5 rounded-full ${status === "running" ? "bg-emerald-500 animate-pulse" : status === "error" ? "bg-rose-500" : "bg-gray-400"}"></span>
                  ${status.charAt(0).toUpperCase() + status.slice(1)}
                </span>
              </div>
            </div>
            ${stats.lastActivity ? html`<span class="text-xs text-gray-500">Last activity: ${new Date(stats.lastActivity).toLocaleTimeString()}</span>` : nothing}
          </div>
          <div class="grid grid-cols-2 gap-4 md:grid-cols-4">
            <button
              type="button"
              class="rounded-md bg-gray-50 px-3 py-2 text-left transition hover:bg-gray-100 cursor-pointer overflow-hidden"
              @click=${() => this.handleMenuSelect("ag-ui-events")}
              title="View all events in AG-UI Events"
            >
              <div class="truncate whitespace-nowrap text-xs text-gray-600">Total Events</div>
              <div class="text-lg font-semibold text-gray-900">${stats.totalEvents}</div>
            </button>
            <div class="rounded-md bg-gray-50 px-3 py-2 overflow-hidden">
              <div class="truncate whitespace-nowrap text-xs text-gray-600">Messages</div>
              <div class="text-lg font-semibold text-gray-900">${stats.messages}</div>
            </div>
            <div class="rounded-md bg-gray-50 px-3 py-2 overflow-hidden">
              <div class="truncate whitespace-nowrap text-xs text-gray-600">Tool Calls</div>
              <div class="text-lg font-semibold text-gray-900">${stats.toolCalls}</div>
            </div>
            <div class="rounded-md bg-gray-50 px-3 py-2 overflow-hidden">
              <div class="truncate whitespace-nowrap text-xs text-gray-600">Errors</div>
              <div class="text-lg font-semibold text-gray-900">${stats.errors}</div>
            </div>
          </div>
        </div>

        <!-- Current State Section -->
        <div class="rounded-lg border border-gray-200 bg-white">
          <div class="border-b border-gray-200 px-4 py-3">
            <h4 class="text-sm font-semibold text-gray-900">Current State</h4>
          </div>
          <div class="overflow-auto p-4">
            ${this.hasRenderableState(state) ? html`
                  <pre class="overflow-auto rounded-md bg-gray-50 p-3 text-xs text-gray-800 max-h-64"><code>${this.formatStateForDisplay(state)}</code></pre>
                ` : html`
                  <div class="flex h-40 items-center justify-center text-xs text-gray-500">
                    <div class="flex items-center gap-2 text-gray-500">
                      <span class="text-lg text-gray-400">${this.renderIcon("Database")}</span>
                      <span>State is empty</span>
                    </div>
                  </div>
                `}
          </div>
        </div>

        <!-- Current Messages Section -->
        <div class="rounded-lg border border-gray-200 bg-white">
          <div class="border-b border-gray-200 px-4 py-3">
            <h4 class="text-sm font-semibold text-gray-900">Current Messages</h4>
          </div>
          <div class="overflow-auto">
            ${messages && messages.length > 0 ? html`
                  <table class="w-full text-xs">
                    <thead class="bg-gray-50">
                      <tr>
                        <th class="px-4 py-2 text-left font-medium text-gray-700">Role</th>
                        <th class="px-4 py-2 text-left font-medium text-gray-700">Content</th>
                      </tr>
                    </thead>
                    <tbody class="divide-y divide-gray-200">
                      ${messages.map((msg) => {
      const role = msg.role || "unknown";
      const roleColors = {
        user: "bg-blue-100 text-blue-800",
        assistant: "bg-green-100 text-green-800",
        system: "bg-gray-100 text-gray-800",
        tool: "bg-amber-100 text-amber-800",
        unknown: "bg-gray-100 text-gray-600"
      };
      const rawContent = msg.contentText ?? "";
      const toolCalls = msg.toolCalls ?? [];
      const hasContent = rawContent.trim().length > 0;
      const contentFallback = toolCalls.length > 0 ? "Invoked tool call" : "\u2014";
      return html`
                          <tr>
                            <td class="px-4 py-2 align-top">
                              <span class="inline-flex rounded px-2 py-0.5 text-[10px] font-medium ${roleColors[role] || roleColors.unknown}">
                                ${role}
                              </span>
                            </td>
                            <td class="px-4 py-2">
                              ${hasContent ? html`<div class="max-w-2xl whitespace-pre-wrap break-words text-gray-700">${rawContent}</div>` : html`<div class="text-xs italic text-gray-400">${contentFallback}</div>`}
                              ${role === "assistant" && toolCalls.length > 0 ? this.renderToolCallDetails(toolCalls) : nothing}
                            </td>
                          </tr>
                        `;
    })}
                    </tbody>
                  </table>
                ` : html`
                  <div class="flex h-40 items-center justify-center text-xs text-gray-500">
                    <div class="flex items-center gap-2 text-gray-500">
                      <span class="text-lg text-gray-400">${this.renderIcon("MessageSquare")}</span>
                      <span>No messages available</span>
                    </div>
                  </div>
                `}
          </div>
        </div>
      </div>
    `;
  }
  renderContextDropdown() {
    const filteredOptions = this.selectedMenu === "agents" ? this.contextOptions.filter((opt) => opt.key !== "all-agents") : this.contextOptions;
    const selectedLabel = filteredOptions.find((opt) => opt.key === this.selectedContext)?.label ?? "";
    return html`
      <div class="relative z-40 min-w-0 flex-1" data-context-dropdown-root="true">
        <button
          type="button"
          class="relative z-40 flex w-full min-w-0 max-w-[240px] items-center gap-1.5 rounded-md border border-gray-200 px-2 py-1 text-xs font-medium text-gray-700 transition hover:border-gray-300 hover:bg-gray-50"
          @pointerdown=${this.handleContextDropdownToggle}
        >
          <span class="truncate flex-1 text-left">${selectedLabel}</span>
          <span class="shrink-0 text-gray-400">${this.renderIcon("ChevronDown")}</span>
        </button>
        ${this.contextMenuOpen ? html`
              <div
                class="absolute left-0 z-50 mt-1.5 w-40 rounded-md border border-gray-200 bg-white py-1 shadow-md ring-1 ring-black/5"
                data-context-dropdown-root="true"
              >
                ${filteredOptions.map(
      (option) => html`
                    <button
                      type="button"
                      class="flex w-full items-center justify-between px-3 py-1.5 text-left text-xs transition hover:bg-gray-50 focus:bg-gray-50 focus:outline-none"
                      data-context-dropdown-root="true"
                      @click=${() => this.handleContextOptionSelect(option.key)}
                    >
                      <span class="truncate ${option.key === this.selectedContext ? "text-gray-900 font-medium" : "text-gray-600"}">${option.label}</span>
                      ${option.key === this.selectedContext ? html`<span class="text-gray-500">${this.renderIcon("Check")}</span>` : nothing}
                    </button>
                  `
    )}
              </div>
            ` : nothing}
      </div>
    `;
  }
  handleMenuSelect(key) {
    if (!this.menuItems.some((item) => item.key === key)) {
      return;
    }
    this.selectedMenu = key;
    if (key === "agents" && this.selectedContext === "all-agents") {
      const agentOptions = this.contextOptions.filter((opt) => opt.key !== "all-agents");
      if (agentOptions.length > 0) {
        const defaultAgent = agentOptions.find((opt) => opt.key === "default");
        this.selectedContext = defaultAgent ? defaultAgent.key : agentOptions[0].key;
      }
    }
    this.contextMenuOpen = false;
    this.persistState();
    this.requestUpdate();
  }
  handleContextDropdownToggle(event) {
    event.preventDefault();
    event.stopPropagation();
    this.contextMenuOpen = !this.contextMenuOpen;
    this.requestUpdate();
  }
  handleContextOptionSelect(key) {
    if (!this.contextOptions.some((option) => option.key === key)) {
      return;
    }
    if (this.selectedContext !== key) {
      this.selectedContext = key;
      this.expandedRows.clear();
    }
    this.contextMenuOpen = false;
    this.persistState();
    this.requestUpdate();
  }
  renderToolsView() {
    if (!this._core) {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-xs text-gray-500">
          No core instance available
        </div>
      `;
    }
    this.refreshToolsSnapshot();
    const allTools = this.cachedTools;
    if (allTools.length === 0) {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-center">
          <div class="max-w-md">
            <div class="mb-3 flex justify-center text-gray-300 [&>svg]:!h-8 [&>svg]:!w-8">
              ${this.renderIcon("Hammer")}
            </div>
            <p class="text-sm text-gray-600">No tools available</p>
            <p class="mt-2 text-xs text-gray-500">Tools will appear here once agents are configured with tool handlers or renderers.</p>
          </div>
        </div>
      `;
    }
    const filteredTools = this.selectedContext === "all-agents" ? allTools : allTools.filter((tool) => !tool.agentId || tool.agentId === this.selectedContext);
    return html`
      <div class="flex h-full flex-col overflow-hidden">
        <div class="overflow-auto p-4">
          <div class="space-y-3">
            ${filteredTools.map((tool) => this.renderToolCard(tool))}
          </div>
        </div>
      </div>
    `;
  }
  extractToolsFromAgents() {
    if (!this._core) {
      return [];
    }
    const tools = [];
    for (const coreTool of this._core.tools ?? []) {
      tools.push({
        agentId: coreTool.agentId ?? "",
        name: coreTool.name,
        description: coreTool.description,
        parameters: coreTool.parameters,
        type: "handler"
      });
    }
    for (const [agentId, agent] of Object.entries(this._core.agents)) {
      if (!agent) continue;
      const handlers = agent.toolHandlers;
      if (handlers && typeof handlers === "object") {
        for (const [toolName, handler] of Object.entries(handlers)) {
          if (handler && typeof handler === "object") {
            const handlerObj = handler;
            tools.push({
              agentId,
              name: toolName,
              description: typeof handlerObj.description === "string" && handlerObj.description || handlerObj.tool?.description,
              parameters: handlerObj.parameters ?? handlerObj.tool?.parameters,
              type: "handler"
            });
          }
        }
      }
      const renderers = agent.toolRenderers;
      if (renderers && typeof renderers === "object") {
        for (const [toolName, renderer] of Object.entries(renderers)) {
          if (!tools.some((t) => t.agentId === agentId && t.name === toolName)) {
            if (renderer && typeof renderer === "object") {
              const rendererObj = renderer;
              tools.push({
                agentId,
                name: toolName,
                description: typeof rendererObj.description === "string" && rendererObj.description || rendererObj.tool?.description,
                parameters: rendererObj.parameters ?? rendererObj.tool?.parameters,
                type: "renderer"
              });
            }
          }
        }
      }
    }
    return tools.sort((a, b) => {
      const agentCompare = a.agentId.localeCompare(b.agentId);
      if (agentCompare !== 0) return agentCompare;
      return a.name.localeCompare(b.name);
    });
  }
  renderToolCard(tool) {
    const isExpanded = this.expandedTools.has(`${tool.agentId}:${tool.name}`);
    const schema = this.extractSchemaInfo(tool.parameters);
    const typeColors = {
      handler: "bg-blue-50 text-blue-700 border-blue-200",
      renderer: "bg-purple-50 text-purple-700 border-purple-200"
    };
    return html`
      <div class="rounded-lg border border-gray-200 bg-white overflow-hidden">
        <button
          type="button"
          class="w-full px-4 py-3 text-left transition hover:bg-gray-50"
          @click=${() => this.toggleToolExpansion(`${tool.agentId}:${tool.name}`)}
        >
          <div class="flex items-start justify-between gap-3">
            <div class="flex-1 min-w-0">
              <div class="flex items-center gap-2 mb-1">
                <span class="font-mono text-sm font-semibold text-gray-900">${tool.name}</span>
                <span class="inline-flex items-center rounded-sm border px-1.5 py-0.5 text-[10px] font-medium ${typeColors[tool.type]}">
                  ${tool.type}
                </span>
              </div>
              <div class="flex items-center gap-2 text-xs text-gray-500">
                <span class="flex items-center gap-1">
                  ${this.renderIcon("Bot")}
                  <span class="font-mono">${tool.agentId}</span>
                </span>
                ${schema.properties.length > 0 ? html`
                      <span class="text-gray-300"></span>
                      <span>${schema.properties.length} parameter${schema.properties.length !== 1 ? "s" : ""}</span>
                    ` : nothing}
              </div>
              ${tool.description ? html`<p class="mt-2 text-xs text-gray-600">${tool.description}</p>` : nothing}
            </div>
            <span class="shrink-0 text-gray-400 transition ${isExpanded ? "rotate-180" : ""}">
              ${this.renderIcon("ChevronDown")}
            </span>
          </div>
        </button>

        ${isExpanded ? html`
              <div class="border-t border-gray-200 bg-gray-50/50 px-4 py-3">
                ${schema.properties.length > 0 ? html`
                      <h5 class="mb-3 text-xs font-semibold text-gray-700">Parameters</h5>
                      <div class="space-y-3">
                        ${schema.properties.map((prop) => html`
                          <div class="rounded-md border border-gray-200 bg-white p-3">
                            <div class="flex items-start justify-between gap-2 mb-1">
                              <span class="font-mono text-xs font-medium text-gray-900">${prop.name}</span>
                              <div class="flex items-center gap-1.5 shrink-0">
                                ${prop.required ? html`<span class="text-[9px] rounded border border-rose-200 bg-rose-50 px-1 py-0.5 font-medium text-rose-700">required</span>` : html`<span class="text-[9px] rounded border border-gray-200 bg-gray-50 px-1 py-0.5 font-medium text-gray-600">optional</span>`}
                                ${prop.type ? html`<span class="text-[9px] rounded border border-gray-200 bg-gray-50 px-1 py-0.5 font-mono text-gray-600">${prop.type}</span>` : nothing}
                              </div>
                            </div>
                            ${prop.description ? html`<p class="mt-1 text-xs text-gray-600">${prop.description}</p>` : nothing}
                            ${prop.defaultValue !== void 0 ? html`
                                  <div class="mt-2 flex items-center gap-1.5 text-[10px] text-gray-500">
                                    <span>Default:</span>
                                    <code class="rounded bg-gray-100 px-1 py-0.5 font-mono">${JSON.stringify(prop.defaultValue)}</code>
                                  </div>
                                ` : nothing}
                            ${prop.enum && prop.enum.length > 0 ? html`
                                  <div class="mt-2">
                                    <span class="text-[10px] text-gray-500">Allowed values:</span>
                                    <div class="mt-1 flex flex-wrap gap-1">
                                      ${prop.enum.map((val) => html`
                                        <code class="rounded border border-gray-200 bg-gray-50 px-1.5 py-0.5 text-[10px] font-mono text-gray-700">${JSON.stringify(val)}</code>
                                      `)}
                                    </div>
                                  </div>
                                ` : nothing}
                          </div>
                        `)}
                      </div>
                    ` : html`
                      <div class="flex items-center justify-center py-4 text-xs text-gray-500">
                        <span>No parameters defined</span>
                      </div>
                    `}
              </div>
            ` : nothing}
      </div>
    `;
  }
  extractSchemaInfo(parameters) {
    const result = { properties: [] };
    if (!parameters || typeof parameters !== "object") {
      return result;
    }
    const zodDef = parameters._def;
    if (zodDef && typeof zodDef === "object") {
      if (zodDef.typeName === "ZodObject") {
        const rawShape = zodDef.shape;
        const shape = typeof rawShape === "function" ? rawShape() : rawShape;
        if (!shape || typeof shape !== "object") {
          return result;
        }
        const requiredKeys = /* @__PURE__ */ new Set();
        if (zodDef.unknownKeys === "strict" || !zodDef.catchall) {
          Object.keys(shape || {}).forEach((key) => {
            const candidate = shape[key];
            const fieldDef = candidate?._def;
            if (fieldDef && !this.isZodOptional(candidate)) {
              requiredKeys.add(key);
            }
          });
        }
        for (const [key, value] of Object.entries(shape || {})) {
          const fieldInfo = this.extractZodFieldInfo(value);
          result.properties.push({
            name: key,
            type: fieldInfo.type,
            description: fieldInfo.description,
            required: requiredKeys.has(key),
            defaultValue: fieldInfo.defaultValue,
            enum: fieldInfo.enum
          });
        }
      }
    } else if (parameters.type === "object" && parameters.properties) {
      const props = parameters.properties;
      const required = new Set(
        Array.isArray(parameters.required) ? parameters.required : []
      );
      for (const [key, value] of Object.entries(props ?? {})) {
        const prop = value;
        result.properties.push({
          name: key,
          type: prop.type,
          description: typeof prop.description === "string" ? prop.description : void 0,
          required: required.has(key),
          defaultValue: prop.default,
          enum: Array.isArray(prop.enum) ? prop.enum : void 0
        });
      }
    }
    return result;
  }
  isZodOptional(zodSchema) {
    const schema = zodSchema;
    if (!schema?._def) return false;
    const def = schema._def;
    if (def.typeName === "ZodOptional" || def.typeName === "ZodNullable") {
      return true;
    }
    if (def.defaultValue !== void 0) {
      return true;
    }
    return false;
  }
  extractZodFieldInfo(zodSchema) {
    const info = {};
    const schema = zodSchema;
    if (!schema?._def) return info;
    let currentSchema = schema;
    let def = currentSchema._def;
    while (def.typeName === "ZodOptional" || def.typeName === "ZodNullable" || def.typeName === "ZodDefault") {
      if (def.typeName === "ZodDefault" && def.defaultValue !== void 0) {
        info.defaultValue = typeof def.defaultValue === "function" ? def.defaultValue() : def.defaultValue;
      }
      currentSchema = def.innerType ?? currentSchema;
      if (!currentSchema?._def) break;
      def = currentSchema._def;
    }
    info.description = typeof def.description === "string" ? def.description : void 0;
    const typeName = typeof def.typeName === "string" ? def.typeName : void 0;
    const typeMap = {
      ZodString: "string",
      ZodNumber: "number",
      ZodBoolean: "boolean",
      ZodArray: "array",
      ZodObject: "object",
      ZodEnum: "enum",
      ZodLiteral: "literal",
      ZodUnion: "union",
      ZodAny: "any",
      ZodUnknown: "unknown"
    };
    info.type = typeName ? typeMap[typeName] || typeName.replace("Zod", "").toLowerCase() : void 0;
    if (typeName === "ZodEnum" && Array.isArray(def.values)) {
      info.enum = def.values;
    } else if (typeName === "ZodLiteral" && def.value !== void 0) {
      info.enum = [def.value];
    }
    return info;
  }
  toggleToolExpansion(toolId) {
    if (this.expandedTools.has(toolId)) {
      this.expandedTools.delete(toolId);
    } else {
      this.expandedTools.add(toolId);
    }
    this.requestUpdate();
  }
  renderContextView() {
    const contextEntries = Object.entries(this.contextStore);
    if (contextEntries.length === 0) {
      return html`
        <div class="flex h-full items-center justify-center px-4 py-8 text-center">
          <div class="max-w-md">
            <div class="mb-3 flex justify-center text-gray-300 [&>svg]:!h-8 [&>svg]:!w-8">
              ${this.renderIcon("FileText")}
            </div>
            <p class="text-sm text-gray-600">No context available</p>
            <p class="mt-2 text-xs text-gray-500">Context will appear here once added to CopilotKit.</p>
          </div>
        </div>
      `;
    }
    return html`
      <div class="flex h-full flex-col overflow-hidden">
        <div class="overflow-auto p-4">
          <div class="space-y-3">
            ${contextEntries.map(([id, context]) => this.renderContextCard(id, context))}
          </div>
        </div>
      </div>
    `;
  }
  renderContextCard(id, context) {
    const isExpanded = this.expandedContextItems.has(id);
    const valuePreview = this.getContextValuePreview(context.value);
    const hasValue = context.value !== void 0 && context.value !== null;
    const title = context.description?.trim() || id;
    return html`
      <div class="rounded-lg border border-gray-200 bg-white overflow-hidden">
        <button
          type="button"
          class="w-full px-4 py-3 text-left transition hover:bg-gray-50"
          @click=${() => this.toggleContextExpansion(id)}
        >
          <div class="flex items-start justify-between gap-3">
            <div class="flex-1 min-w-0">
              <p class="text-sm font-medium text-gray-900 mb-1">${title}</p>
              <div class="flex items-center gap-2 text-xs text-gray-500">
                <span class="font-mono truncate inline-block align-middle" style="max-width: 180px;">${id}</span>
                ${hasValue ? html`
                      <span class="text-gray-300"></span>
                      <span class="truncate">${valuePreview}</span>
                    ` : nothing}
              </div>
            </div>
            <span class="shrink-0 text-gray-400 transition ${isExpanded ? "rotate-180" : ""}">
              ${this.renderIcon("ChevronDown")}
            </span>
          </div>
        </button>

        ${isExpanded ? html`
              <div class="border-t border-gray-200 bg-gray-50/50 px-4 py-3">
                <div class="mb-3">
                  <h5 class="mb-1 text-xs font-semibold text-gray-700">ID</h5>
                  <code class="block rounded bg-white border border-gray-200 px-2 py-1 text-[10px] font-mono text-gray-600">${id}</code>
                </div>
                ${hasValue ? html`
                      <div class="mb-2 flex items-center justify-between gap-2">
                        <h5 class="text-xs font-semibold text-gray-700">Value</h5>
                        <button
                          class="flex items-center gap-1 rounded-md border border-gray-200 bg-white px-2 py-1 text-[10px] font-medium text-gray-700 transition hover:bg-gray-50"
                          type="button"
                          @click=${(e) => {
      e.stopPropagation();
      void this.copyContextValue(context.value, id);
    }}
                        >
                          ${this.copiedContextItems.has(id) ? "Copied" : "Copy JSON"}
                        </button>
                      </div>
                      <div class="rounded-md border border-gray-200 bg-white p-3">
                        <pre class="overflow-auto text-xs text-gray-800 max-h-96"><code>${this.formatContextValue(context.value)}</code></pre>
                      </div>
                    ` : html`
                      <div class="flex items-center justify-center py-4 text-xs text-gray-500">
                        <span>No value available</span>
                      </div>
                    `}
              </div>
            ` : nothing}
      </div>
    `;
  }
  getContextValuePreview(value) {
    if (value === void 0 || value === null) {
      return "\u2014";
    }
    if (typeof value === "string") {
      return value.length > 50 ? `${value.substring(0, 50)}...` : value;
    }
    if (typeof value === "number" || typeof value === "boolean") {
      return String(value);
    }
    if (Array.isArray(value)) {
      return `Array(${value.length})`;
    }
    if (typeof value === "object") {
      const keys = Object.keys(value);
      return `Object with ${keys.length} key${keys.length !== 1 ? "s" : ""}`;
    }
    if (typeof value === "function") {
      return "Function";
    }
    return String(value);
  }
  formatContextValue(value) {
    if (value === void 0) {
      return "undefined";
    }
    if (value === null) {
      return "null";
    }
    if (typeof value === "function") {
      return value.toString();
    }
    try {
      return JSON.stringify(value, null, 2);
    } catch {
      return String(value);
    }
  }
  async copyContextValue(value, contextId) {
    if (typeof navigator === "undefined" || !navigator.clipboard?.writeText) {
      console.warn("Clipboard API is not available in this environment.");
      return;
    }
    const serialized = this.formatContextValue(value);
    try {
      await navigator.clipboard.writeText(serialized);
      this.copiedContextItems.add(contextId);
      this.requestUpdate();
      setTimeout(() => {
        this.copiedContextItems.delete(contextId);
        this.requestUpdate();
      }, 1500);
    } catch (error) {
      console.error("Failed to copy context value:", error);
    }
  }
  toggleContextExpansion(contextId) {
    if (this.expandedContextItems.has(contextId)) {
      this.expandedContextItems.delete(contextId);
    } else {
      this.expandedContextItems.add(contextId);
    }
    this.requestUpdate();
  }
  toggleRowExpansion(eventId) {
    const selection = window.getSelection();
    if (selection && selection.toString().length > 0) {
      return;
    }
    if (this.expandedRows.has(eventId)) {
      this.expandedRows.delete(eventId);
    } else {
      this.expandedRows.add(eventId);
    }
    this.requestUpdate();
  }
  renderAnnouncementPanel() {
    if (!this.isOpen) {
      return nothing;
    }
    this.ensureAnnouncementLoading();
    if (!this.hasUnseenAnnouncement) {
      return nothing;
    }
    if (!this.announcementLoaded && !this.announcementMarkdown) {
      return html`<div class="mx-4 my-3 rounded-xl border border-slate-200 bg-white px-4 py-3 text-sm text-slate-800 shadow-[0_12px_30px_rgba(15,23,42,0.12)]">
        <div class="flex items-center gap-2 font-semibold">
          <span class="inline-flex h-6 w-6 items-center justify-center rounded-md bg-slate-900 text-white shadow-sm">
            ${this.renderIcon("Megaphone")}
          </span>
          <span>Loading latest announcement</span>
        </div>
      </div>`;
    }
    if (this.announcementLoadError) {
      return html`<div class="mx-4 my-3 rounded-xl border border-rose-200 bg-rose-50 px-4 py-3 text-sm text-rose-900 shadow-[0_12px_30px_rgba(15,23,42,0.12)]">
        <div class="flex items-center gap-2 font-semibold">
          <span class="inline-flex h-6 w-6 items-center justify-center rounded-md bg-rose-600 text-white shadow-sm">
            ${this.renderIcon("Megaphone")}
          </span>
          <span>Announcement unavailable</span>
        </div>
        <p class="mt-2 text-xs text-rose-800">We couldnt load the latest notice. Please try opening the inspector again.</p>
      </div>`;
    }
    if (!this.announcementMarkdown) {
      return nothing;
    }
    const content = this.announcementHtml ? unsafeHTML(this.announcementHtml) : html`<pre class="whitespace-pre-wrap text-sm text-gray-900">${this.announcementMarkdown}</pre>`;
    return html`<div class="mx-4 my-3 rounded-xl border border-slate-200 bg-white px-4 py-4 shadow-[0_12px_30px_rgba(15,23,42,0.12)]">
      <div class="mb-3 flex items-center gap-2 text-sm font-semibold text-slate-900">
        <span class="inline-flex h-7 w-7 items-center justify-center rounded-md bg-slate-900 text-white shadow-sm">
          ${this.renderIcon("Megaphone")}
        </span>
        <span>Announcement</span>
        <button class="announcement-dismiss ml-auto" type="button" @click=${this.handleDismissAnnouncement} aria-label="Dismiss announcement">
          Dismiss
        </button>
      </div>
      <div class="announcement-content text-sm leading-relaxed text-gray-900">${content}</div>
    </div>`;
  }
  ensureAnnouncementLoading() {
    if (this.announcementPromise || typeof window === "undefined" || typeof fetch === "undefined") {
      return;
    }
    this.announcementPromise = this.fetchAnnouncement();
  }
  renderAnnouncementPreview() {
    if (!this.hasUnseenAnnouncement || !this.showAnnouncementPreview || !this.announcementPreviewText) {
      return nothing;
    }
    const side = this.contextState.button.anchor.horizontal === "left" ? "right" : "left";
    return html`<div
      class="announcement-preview"
      data-side=${side}
      role="note"
      @click=${() => this.handleAnnouncementPreviewClick()}
    >
      <span>${this.announcementPreviewText}</span>
      <span class="announcement-preview__arrow"></span>
    </div>`;
  }
  handleAnnouncementPreviewClick() {
    this.showAnnouncementPreview = false;
    this.openInspector();
  }
  async fetchAnnouncement() {
    try {
      const response = await fetch(ANNOUNCEMENT_URL, { cache: "no-cache" });
      if (!response.ok) {
        throw new Error(`Failed to load announcement (${response.status})`);
      }
      const data = await response.json();
      const timestamp = typeof data?.timestamp === "string" ? data.timestamp : null;
      const previewText = typeof data?.previewText === "string" ? data.previewText : null;
      const markdown = typeof data?.announcement === "string" ? data.announcement : null;
      if (!timestamp || !markdown) {
        throw new Error("Malformed announcement payload");
      }
      const storedTimestamp = this.loadStoredAnnouncementTimestamp();
      this.announcementTimestamp = timestamp;
      this.announcementPreviewText = previewText ?? "";
      this.announcementMarkdown = markdown;
      this.hasUnseenAnnouncement = (!storedTimestamp || storedTimestamp !== timestamp) && !!this.announcementPreviewText;
      this.showAnnouncementPreview = this.hasUnseenAnnouncement;
      this.announcementHtml = await this.convertMarkdownToHtml(markdown);
      this.announcementLoaded = true;
      this.requestUpdate();
    } catch (error) {
      this.announcementLoadError = error;
      this.announcementLoaded = true;
      this.requestUpdate();
    }
  }
  async convertMarkdownToHtml(markdown) {
    const renderer = new marked.Renderer();
    renderer.link = (href, title, text) => {
      const safeHref = this.escapeHtmlAttr(this.appendRefParam(href ?? ""));
      const titleAttr = title ? ` title="${this.escapeHtmlAttr(title)}"` : "";
      return `<a href="${safeHref}" target="_blank" rel="noopener"${titleAttr}>${text}</a>`;
    };
    return marked.parse(markdown, { renderer });
  }
  appendRefParam(href) {
    try {
      const url = new URL(href, typeof window !== "undefined" ? window.location.href : "https://copilotkit.ai");
      if (!url.searchParams.has("ref")) {
        url.searchParams.append("ref", "cpk-inspector");
      }
      return url.toString();
    } catch {
      return href;
    }
  }
  escapeHtmlAttr(value) {
    return value.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/\"/g, "&quot;").replace(/'/g, "&#39;");
  }
  loadStoredAnnouncementTimestamp() {
    if (typeof window === "undefined" || !window.localStorage) {
      return null;
    }
    try {
      const raw = window.localStorage.getItem(ANNOUNCEMENT_STORAGE_KEY);
      if (!raw) {
        return null;
      }
      const parsed = JSON.parse(raw);
      if (parsed && typeof parsed.timestamp === "string") {
        return parsed.timestamp;
      }
      return null;
    } catch {
    }
    return null;
  }
  persistAnnouncementTimestamp(timestamp) {
    if (typeof window === "undefined" || !window.localStorage) {
      return;
    }
    try {
      const payload = JSON.stringify({ timestamp });
      window.localStorage.setItem(ANNOUNCEMENT_STORAGE_KEY, payload);
    } catch {
    }
  }
  markAnnouncementSeen() {
    this.hasUnseenAnnouncement = false;
    this.showAnnouncementPreview = false;
    if (!this.announcementTimestamp) {
      if (this.announcementPromise && !this.announcementLoaded) {
        void this.announcementPromise.then(() => this.markAnnouncementSeen()).catch(() => void 0);
      }
      this.requestUpdate();
      return;
    }
    this.persistAnnouncementTimestamp(this.announcementTimestamp);
    this.requestUpdate();
  }
};
function defineWebInspector() {
  if (!customElements.get(WEB_INSPECTOR_TAG)) {
    customElements.define(WEB_INSPECTOR_TAG, WebInspectorElement);
  }
}
defineWebInspector();
export {
  WEB_INSPECTOR_TAG,
  WebInspectorElement,
  defineWebInspector
};
