// @copilotkit/react-core/dist/hooks/use-copilot-action.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/hooks/use-copilot-action.ts
var use_copilot_action_exports = {};
__export(use_copilot_action_exports, {
  useCopilotAction: () => useCopilotAction
});
module.exports = __toCommonJS(use_copilot_action_exports);
var import_react7 = require("react");

// src/hooks/use-frontend-tool.ts
var import_react = __toESM(require("react"));
var import_shared = require("@copilotkit/shared");
var import_shared2 = require("@copilotkit/shared");
var import_react2 = require("@copilotkitnext/react");
function useFrontendTool(tool, dependencies) {
  const { name, description, parameters, render, followUp } = tool;
  const zodParameters = (0, import_shared.getZodParameters)(parameters);
  const renderRef = (0, import_react.useRef)(render);
  (0, import_react.useEffect)(() => {
    renderRef.current = render;
  }, [render, ...dependencies != null ? dependencies : []]);
  const normalizedRender = (0, import_react.useMemo)(() => {
    if (typeof render === "undefined") {
      return void 0;
    }
    return (args) => {
      const currentRender = renderRef.current;
      if (typeof currentRender === "undefined") {
        return null;
      }
      if (typeof currentRender === "string") {
        return import_react.default.createElement(import_react.default.Fragment, null, currentRender);
      }
      const renderArgs = __spreadProps(__spreadValues({}, args), {
        result: typeof args.result === "string" ? (0, import_shared2.parseJson)(args.result, args.result) : args.result
      });
      const rendered = currentRender(renderArgs);
      if (typeof rendered === "string") {
        return import_react.default.createElement(import_react.default.Fragment, null, rendered);
      }
      return rendered != null ? rendered : null;
    };
  }, []);
  const handlerRef = (0, import_react.useRef)(tool.handler);
  (0, import_react.useEffect)(() => {
    handlerRef.current = tool.handler;
  }, [tool.handler, ...dependencies != null ? dependencies : []]);
  const normalizedHandler = tool.handler ? (args) => {
    var _a;
    return (_a = handlerRef.current) == null ? void 0 : _a.call(handlerRef, args);
  } : void 0;
  (0, import_react2.useFrontendTool)({
    name,
    description,
    parameters: zodParameters,
    handler: normalizedHandler,
    followUp,
    render: normalizedRender
  });
}

// src/hooks/use-render-tool-call.ts
var import_shared3 = require("@copilotkit/shared");
var import_react3 = require("react");
var import_react4 = require("@copilotkitnext/react");
var import_shared4 = require("@copilotkit/shared");
function useRenderToolCall(tool, dependencies) {
  const { copilotkit } = (0, import_react4.useCopilotKit)();
  const hasAddedRef = (0, import_react3.useRef)(false);
  (0, import_react3.useEffect)(() => {
    const { name, parameters, render } = tool;
    const zodParameters = (0, import_shared3.getZodParameters)(parameters);
    const renderToolCall = name === "*" ? (0, import_react4.defineToolCallRenderer)({
      name: "*",
      render: (args) => {
        return render(__spreadProps(__spreadValues({}, args), {
          result: args.result ? (0, import_shared4.parseJson)(args.result, args.result) : args.result
        }));
      }
    }) : (0, import_react4.defineToolCallRenderer)({
      name,
      args: zodParameters,
      render: (args) => {
        return render(__spreadProps(__spreadValues({}, args), {
          result: args.result ? (0, import_shared4.parseJson)(args.result, args.result) : args.result
        }));
      }
    });
    const existingIndex = copilotkit.renderToolCalls.findIndex((r) => r.name === name);
    if (existingIndex !== -1) {
      copilotkit.renderToolCalls.splice(existingIndex, 1);
    }
    copilotkit.renderToolCalls.push(renderToolCall);
    hasAddedRef.current = true;
    return () => {
      if (hasAddedRef.current) {
        const index = copilotkit.renderToolCalls.findIndex((r) => r.name === name);
        if (index !== -1) {
          copilotkit.renderToolCalls.splice(index, 1);
        }
        hasAddedRef.current = false;
      }
    };
  }, [tool, ...dependencies != null ? dependencies : []]);
}

// src/hooks/use-human-in-the-loop.ts
var import_shared5 = require("@copilotkit/shared");
var import_react5 = require("@copilotkitnext/react");
var import_core = require("@copilotkitnext/core");
var import_react6 = __toESM(require("react"));
function useHumanInTheLoop(tool, dependencies) {
  const _a = tool, { render } = _a, toolRest = __objRest(_a, ["render"]);
  const { name, description, parameters, followUp } = toolRest;
  const zodParameters = (0, import_shared5.getZodParameters)(parameters);
  const renderRef = (0, import_react6.useRef)(null);
  (0, import_react6.useEffect)(() => {
    renderRef.current = (args) => {
      if (typeof render === "string") {
        return import_react6.default.createElement(import_react6.default.Fragment, null, render);
      }
      if (!render) {
        return null;
      }
      const renderProps = (() => {
        const mappedArgs = args.args;
        switch (args.status) {
          case import_core.ToolCallStatus.InProgress:
            return {
              args: mappedArgs,
              respond: args.respond,
              status: args.status,
              handler: void 0
            };
          case import_core.ToolCallStatus.Executing:
            return {
              args: mappedArgs,
              respond: args.respond,
              status: args.status,
              handler: () => {
              }
            };
          case import_core.ToolCallStatus.Complete:
            return {
              args: mappedArgs,
              respond: args.respond,
              status: args.status,
              result: args.result ? (0, import_shared5.parseJson)(args.result, args.result) : args.result,
              handler: void 0
            };
          default:
            throw new import_shared5.CopilotKitError({
              code: import_shared5.CopilotKitErrorCode.UNKNOWN,
              message: `Invalid tool call status: ${args.status}`
            });
        }
      })();
      const rendered = render(renderProps);
      if (typeof rendered === "string") {
        return import_react6.default.createElement(import_react6.default.Fragment, null, rendered);
      }
      return rendered != null ? rendered : null;
    };
  }, [render, ...dependencies != null ? dependencies : []]);
  (0, import_react5.useHumanInTheLoop)({
    name,
    description,
    followUp,
    parameters: zodParameters,
    render: (args) => {
      var _a2, _b;
      return (_b = (_a2 = renderRef.current) == null ? void 0 : _a2.call(renderRef, args)) != null ? _b : null;
    }
  });
}

// src/hooks/use-copilot-action.ts
function getActionConfig(action) {
  if (action.name === "*") {
    return {
      type: "render",
      action
    };
  }
  if ("renderAndWaitForResponse" in action || "renderAndWait" in action) {
    let render = action.render;
    if (!render && "renderAndWaitForResponse" in action) {
      render = action.renderAndWaitForResponse;
    }
    if (!render && "renderAndWait" in action) {
      render = action.renderAndWait;
    }
    return {
      type: "hitl",
      action: __spreadProps(__spreadValues({}, action), { render })
    };
  }
  if ("available" in action) {
    if (action.available === "enabled" || action.available === "remote") {
      return {
        type: "frontend",
        action
      };
    }
    if (action.available === "frontend" || action.available === "disabled") {
      return {
        type: "render",
        action
      };
    }
  }
  if ("handler" in action) {
    return {
      type: "frontend",
      action
    };
  }
  throw new Error("Invalid action configuration");
}
function useCopilotAction(action, dependencies) {
  const [initialActionConfig] = (0, import_react7.useState)(getActionConfig(action));
  const currentActionConfig = getActionConfig(action);
  if (initialActionConfig.type !== currentActionConfig.type) {
    throw new Error("Action configuration changed between renders");
  }
  switch (currentActionConfig.type) {
    case "render":
      return useRenderToolCall(currentActionConfig.action, dependencies);
    case "hitl":
      return useHumanInTheLoop(currentActionConfig.action, dependencies);
    case "frontend":
      return useFrontendTool(currentActionConfig.action, dependencies);
    default:
      throw new Error("Invalid action configuration");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useCopilotAction
});
