// @copilotkit/react-ui/dist/index.mjs
"use client";
import "./chunk-EFZPSZWO.mjs";
import "./chunk-MMVDU6DF.mjs";
import "./chunk-SC6JRFAJ.mjs";
import "./chunk-WB3YULQ4.mjs";
import {
  CopilotPopup
} from "./chunk-7CMZ7YPQ.mjs";
import {
  CopilotSidebar
} from "./chunk-RKQM2AE5.mjs";
import "./chunk-OHGUQZJT.mjs";
import "./chunk-C3GSYRC3.mjs";
import "./chunk-GDSZGYCE.mjs";
import "./chunk-V7W6IM2V.mjs";
import {
  CopilotDevConsole
} from "./chunk-FFJHOZX6.mjs";
import "./chunk-Q5V6S67N.mjs";
import {
  shouldShowDevConsole
} from "./chunk-JY2CSDKN.mjs";
import "./chunk-KXE2JCUH.mjs";
import "./chunk-NRA3CFEE.mjs";
import "./chunk-BH6PCAAL.mjs";
import "./chunk-UFN2VWSR.mjs";
import {
  CopilotChat
} from "./chunk-PD36EMOZ.mjs";
import {
  Suggestions
} from "./chunk-VHKHX7GV.mjs";
import {
  Suggestion
} from "./chunk-6TWQUA2Q.mjs";
import "./chunk-PLHTVHUW.mjs";
import "./chunk-3W6J75HS.mjs";
import "./chunk-ELGRNEAO.mjs";
import "./chunk-QIOJXTIQ.mjs";
import "./chunk-WNX755OE.mjs";
import "./chunk-F2CDEOFX.mjs";
import "./chunk-LBR23XXF.mjs";
import {
  AssistantMessage
} from "./chunk-LYYDXL2G.mjs";
import {
  ImageRenderer
} from "./chunk-DBKRAOH7.mjs";
import {
  UserMessage
} from "./chunk-RYUCX3ZK.mjs";
import {
  Markdown
} from "./chunk-JZ3RFQQ6.mjs";
import "./chunk-IK2BPURM.mjs";
import {
  useChatContext
} from "./chunk-IEMQ2SQW.mjs";
import "./chunk-XWG3L6QC.mjs";
import "./chunk-IU3WTXLQ.mjs";
import "./chunk-T26KLXLH.mjs";
import {
  useCopilotChatSuggestions
} from "./chunk-Y4FKRAKJ.mjs";
import "./chunk-54JAUBUJ.mjs";
import "./chunk-JGMFJZMG.mjs";
import "./chunk-O72ZB5V3.mjs";
import "./chunk-MRXNTQOX.mjs";
export {
  AssistantMessage,
  CopilotChat,
  CopilotDevConsole,
  CopilotPopup,
  CopilotSidebar,
  ImageRenderer,
  Markdown,
  Suggestion as RenderSuggestion,
  Suggestions as RenderSuggestionsList,
  UserMessage,
  shouldShowDevConsole,
  useChatContext,
  useCopilotChatSuggestions
};
