// @copilotkit/runtime-client-gql/dist/message-conversion/index.mjs
import "../chunk-CA4VMP2C.mjs";
import {
  aguiMessageWithImageToGQLMessage,
  aguiMessageWithRenderToGQL,
  aguiTextMessageToGQLMessage,
  aguiToGQL,
  aguiToolCallToGQLActionExecution,
  aguiToolMessageToGQLResultMessage
} from "../chunk-RNENVFJ3.mjs";
import {
  gqlActionExecutionMessageToAGUIMessage,
  gqlImageMessageToAGUIMessage,
  gqlResultMessageToAGUIMessage,
  gqlTextMessageToAGUIMessage,
  gqlToAGUI
} from "../chunk-EIPAWDCN.mjs";
import "../chunk-7ECCT6PK.mjs";
import "../chunk-QAHJHPE2.mjs";
import "../chunk-3ETRQN2E.mjs";
import "../chunk-KTA7QTTU.mjs";
import "../chunk-M5LBGBWO.mjs";
import "../chunk-AMFKR5ST.mjs";
import "../chunk-NHOKFX55.mjs";
import "../chunk-OBDQ5BWD.mjs";
import "../chunk-DELDZXUX.mjs";
export {
  aguiMessageWithImageToGQLMessage,
  aguiMessageWithRenderToGQL,
  aguiTextMessageToGQLMessage,
  aguiToGQL,
  aguiToolCallToGQLActionExecution,
  aguiToolMessageToGQLResultMessage,
  gqlActionExecutionMessageToAGUIMessage,
  gqlImageMessageToAGUIMessage,
  gqlResultMessageToAGUIMessage,
  gqlTextMessageToAGUIMessage,
  gqlToAGUI
};
